{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "import os\n",
    "import glob\n",
    "import snapatac2 as snap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "def merge_region(region='PFC',condition=\"MC\",folder='/data2st1/junyi/output/atac0416/dar/region_nt/',method='mementob',blacklist=['Doublet','NN']):\n",
    "    df_result = pd.DataFrame()\n",
    "    files = glob.glob(f'{folder}/{region}*{condition}_{method}.csv')\n",
    "    for file in files:\n",
    "        if any([file.find(black) > -1 for black in blacklist]):\n",
    "            continue\n",
    "        df_dar = pd.read_csv(file,index_col=0)\n",
    "\n",
    "        ctname = file.split('/')[-1].replace('.csv', '')\n",
    "        ctname = \"_\".join(ctname.split('_')[1:-2])\n",
    "        if len(df_dar) == 0:\n",
    "            continue\n",
    "        df_dar['ctname'] = ctname\n",
    "        df_result = pd.concat([df_result, df_dar], axis=0, ignore_index=True)\n",
    "    df_result['region'] = region\n",
    "    df_result['condition'] = condition\n",
    "    return df_result\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_PFC_MC = merge_region(region='PFC',condition=\"MC\")\n",
    "df_PFC_MW = merge_region(region='PFC',condition=\"MW\")\n",
    "df_HIP_MC = merge_region(region='HIP',condition=\"MC\")\n",
    "df_HIP_MW = merge_region(region='HIP',condition=\"MW\")\n",
    "df_AMY_MC = merge_region(region='AMY',condition=\"MC\")\n",
    "df_AMY_MW = merge_region(region='AMY',condition=\"MW\")\n",
    "df_DAR_ALL = pd.concat([df_PFC_MC, df_PFC_MW, df_HIP_MC, df_HIP_MW, df_AMY_MC, df_AMY_MW], axis=0, ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>tx</th>\n",
       "      <th>de_coef</th>\n",
       "      <th>de_se</th>\n",
       "      <th>de_pval</th>\n",
       "      <th>dv_coef</th>\n",
       "      <th>dv_se</th>\n",
       "      <th>dv_pval</th>\n",
       "      <th>ctname</th>\n",
       "      <th>region</th>\n",
       "      <th>condition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr15:75086569-75087070</td>\n",
       "      <td>stim</td>\n",
       "      <td>0.419844</td>\n",
       "      <td>0.017950</td>\n",
       "      <td>2.906623e-121</td>\n",
       "      <td>-0.261757</td>\n",
       "      <td>0.092274</td>\n",
       "      <td>5.066288e-03</td>\n",
       "      <td>PFC_Glut</td>\n",
       "      <td>PFC</td>\n",
       "      <td>MC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr9:76305059-76305560</td>\n",
       "      <td>stim</td>\n",
       "      <td>0.682124</td>\n",
       "      <td>0.031764</td>\n",
       "      <td>6.342578e-102</td>\n",
       "      <td>-0.202397</td>\n",
       "      <td>0.088147</td>\n",
       "      <td>2.124100e-02</td>\n",
       "      <td>PFC_Glut</td>\n",
       "      <td>PFC</td>\n",
       "      <td>MC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX:143482812-143483313</td>\n",
       "      <td>stim</td>\n",
       "      <td>0.610629</td>\n",
       "      <td>0.028551</td>\n",
       "      <td>8.848753e-102</td>\n",
       "      <td>-0.380780</td>\n",
       "      <td>0.125812</td>\n",
       "      <td>3.853716e-03</td>\n",
       "      <td>PFC_Glut</td>\n",
       "      <td>PFC</td>\n",
       "      <td>MC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr12:120028551-120029052</td>\n",
       "      <td>stim</td>\n",
       "      <td>0.397639</td>\n",
       "      <td>0.019140</td>\n",
       "      <td>3.674064e-96</td>\n",
       "      <td>0.007421</td>\n",
       "      <td>0.082149</td>\n",
       "      <td>9.290066e-01</td>\n",
       "      <td>PFC_Glut</td>\n",
       "      <td>PFC</td>\n",
       "      <td>MC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr2:98662045-98662546</td>\n",
       "      <td>stim</td>\n",
       "      <td>0.469970</td>\n",
       "      <td>0.024888</td>\n",
       "      <td>7.864128e-80</td>\n",
       "      <td>-0.503358</td>\n",
       "      <td>0.091436</td>\n",
       "      <td>7.121638e-08</td>\n",
       "      <td>PFC_Glut</td>\n",
       "      <td>PFC</td>\n",
       "      <td>MC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120138</th>\n",
       "      <td>chr1:167001277-167001778</td>\n",
       "      <td>stim</td>\n",
       "      <td>-0.073560</td>\n",
       "      <td>0.037770</td>\n",
       "      <td>4.972503e-02</td>\n",
       "      <td>-0.140123</td>\n",
       "      <td>0.147708</td>\n",
       "      <td>3.389143e-01</td>\n",
       "      <td>AMY_GABA</td>\n",
       "      <td>AMY</td>\n",
       "      <td>MW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120139</th>\n",
       "      <td>chr18:88894092-88894593</td>\n",
       "      <td>stim</td>\n",
       "      <td>-0.055283</td>\n",
       "      <td>0.028996</td>\n",
       "      <td>4.983414e-02</td>\n",
       "      <td>-0.000488</td>\n",
       "      <td>0.100180</td>\n",
       "      <td>9.568052e-01</td>\n",
       "      <td>AMY_GABA</td>\n",
       "      <td>AMY</td>\n",
       "      <td>MW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120140</th>\n",
       "      <td>chr13:110741629-110742130</td>\n",
       "      <td>stim</td>\n",
       "      <td>-0.066564</td>\n",
       "      <td>0.034950</td>\n",
       "      <td>4.985318e-02</td>\n",
       "      <td>0.142614</td>\n",
       "      <td>0.091184</td>\n",
       "      <td>1.316118e-01</td>\n",
       "      <td>AMY_GABA</td>\n",
       "      <td>AMY</td>\n",
       "      <td>MW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120141</th>\n",
       "      <td>chr19:23980544-23981045</td>\n",
       "      <td>stim</td>\n",
       "      <td>-0.068154</td>\n",
       "      <td>0.035806</td>\n",
       "      <td>4.986629e-02</td>\n",
       "      <td>0.000641</td>\n",
       "      <td>0.086747</td>\n",
       "      <td>9.816067e-01</td>\n",
       "      <td>AMY_GABA</td>\n",
       "      <td>AMY</td>\n",
       "      <td>MW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120142</th>\n",
       "      <td>chr3:7366363-7366864</td>\n",
       "      <td>stim</td>\n",
       "      <td>-0.043569</td>\n",
       "      <td>0.022436</td>\n",
       "      <td>4.990000e-02</td>\n",
       "      <td>-0.036747</td>\n",
       "      <td>0.083953</td>\n",
       "      <td>6.686464e-01</td>\n",
       "      <td>AMY_GABA</td>\n",
       "      <td>AMY</td>\n",
       "      <td>MW</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>120143 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             gene    tx   de_coef     de_se        de_pval  \\\n",
       "0         chr15:75086569-75087070  stim  0.419844  0.017950  2.906623e-121   \n",
       "1          chr9:76305059-76305560  stim  0.682124  0.031764  6.342578e-102   \n",
       "2        chrX:143482812-143483313  stim  0.610629  0.028551  8.848753e-102   \n",
       "3       chr12:120028551-120029052  stim  0.397639  0.019140   3.674064e-96   \n",
       "4          chr2:98662045-98662546  stim  0.469970  0.024888   7.864128e-80   \n",
       "...                           ...   ...       ...       ...            ...   \n",
       "120138   chr1:167001277-167001778  stim -0.073560  0.037770   4.972503e-02   \n",
       "120139    chr18:88894092-88894593  stim -0.055283  0.028996   4.983414e-02   \n",
       "120140  chr13:110741629-110742130  stim -0.066564  0.034950   4.985318e-02   \n",
       "120141    chr19:23980544-23981045  stim -0.068154  0.035806   4.986629e-02   \n",
       "120142       chr3:7366363-7366864  stim -0.043569  0.022436   4.990000e-02   \n",
       "\n",
       "         dv_coef     dv_se       dv_pval    ctname region condition  \n",
       "0      -0.261757  0.092274  5.066288e-03  PFC_Glut    PFC        MC  \n",
       "1      -0.202397  0.088147  2.124100e-02  PFC_Glut    PFC        MC  \n",
       "2      -0.380780  0.125812  3.853716e-03  PFC_Glut    PFC        MC  \n",
       "3       0.007421  0.082149  9.290066e-01  PFC_Glut    PFC        MC  \n",
       "4      -0.503358  0.091436  7.121638e-08  PFC_Glut    PFC        MC  \n",
       "...          ...       ...           ...       ...    ...       ...  \n",
       "120138 -0.140123  0.147708  3.389143e-01  AMY_GABA    AMY        MW  \n",
       "120139 -0.000488  0.100180  9.568052e-01  AMY_GABA    AMY        MW  \n",
       "120140  0.142614  0.091184  1.316118e-01  AMY_GABA    AMY        MW  \n",
       "120141  0.000641  0.086747  9.816067e-01  AMY_GABA    AMY        MW  \n",
       "120142 -0.036747  0.083953  6.686464e-01  AMY_GABA    AMY        MW  \n",
       "\n",
       "[120143 rows x 11 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_DAR_ALL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>tx</th>\n",
       "      <th>de_coef</th>\n",
       "      <th>de_se</th>\n",
       "      <th>de_pval</th>\n",
       "      <th>dv_coef</th>\n",
       "      <th>dv_se</th>\n",
       "      <th>dv_pval</th>\n",
       "      <th>ctname</th>\n",
       "      <th>region</th>\n",
       "      <th>condition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr15:75086569-75087070</td>\n",
       "      <td>stim</td>\n",
       "      <td>0.419844</td>\n",
       "      <td>0.017950</td>\n",
       "      <td>2.906623e-121</td>\n",
       "      <td>-0.261757</td>\n",
       "      <td>0.092274</td>\n",
       "      <td>5.066288e-03</td>\n",
       "      <td>PFC_Glut</td>\n",
       "      <td>PFC</td>\n",
       "      <td>MC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr9:76305059-76305560</td>\n",
       "      <td>stim</td>\n",
       "      <td>0.682124</td>\n",
       "      <td>0.031764</td>\n",
       "      <td>6.342578e-102</td>\n",
       "      <td>-0.202397</td>\n",
       "      <td>0.088147</td>\n",
       "      <td>2.124100e-02</td>\n",
       "      <td>PFC_Glut</td>\n",
       "      <td>PFC</td>\n",
       "      <td>MC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX:143482812-143483313</td>\n",
       "      <td>stim</td>\n",
       "      <td>0.610629</td>\n",
       "      <td>0.028551</td>\n",
       "      <td>8.848753e-102</td>\n",
       "      <td>-0.380780</td>\n",
       "      <td>0.125812</td>\n",
       "      <td>3.853716e-03</td>\n",
       "      <td>PFC_Glut</td>\n",
       "      <td>PFC</td>\n",
       "      <td>MC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr12:120028551-120029052</td>\n",
       "      <td>stim</td>\n",
       "      <td>0.397639</td>\n",
       "      <td>0.019140</td>\n",
       "      <td>3.674064e-96</td>\n",
       "      <td>0.007421</td>\n",
       "      <td>0.082149</td>\n",
       "      <td>9.290066e-01</td>\n",
       "      <td>PFC_Glut</td>\n",
       "      <td>PFC</td>\n",
       "      <td>MC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr2:98662045-98662546</td>\n",
       "      <td>stim</td>\n",
       "      <td>0.469970</td>\n",
       "      <td>0.024888</td>\n",
       "      <td>7.864128e-80</td>\n",
       "      <td>-0.503358</td>\n",
       "      <td>0.091436</td>\n",
       "      <td>7.121638e-08</td>\n",
       "      <td>PFC_Glut</td>\n",
       "      <td>PFC</td>\n",
       "      <td>MC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120138</th>\n",
       "      <td>chr1:167001277-167001778</td>\n",
       "      <td>stim</td>\n",
       "      <td>-0.073560</td>\n",
       "      <td>0.037770</td>\n",
       "      <td>4.972503e-02</td>\n",
       "      <td>-0.140123</td>\n",
       "      <td>0.147708</td>\n",
       "      <td>3.389143e-01</td>\n",
       "      <td>AMY_GABA</td>\n",
       "      <td>AMY</td>\n",
       "      <td>MW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120139</th>\n",
       "      <td>chr18:88894092-88894593</td>\n",
       "      <td>stim</td>\n",
       "      <td>-0.055283</td>\n",
       "      <td>0.028996</td>\n",
       "      <td>4.983414e-02</td>\n",
       "      <td>-0.000488</td>\n",
       "      <td>0.100180</td>\n",
       "      <td>9.568052e-01</td>\n",
       "      <td>AMY_GABA</td>\n",
       "      <td>AMY</td>\n",
       "      <td>MW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120140</th>\n",
       "      <td>chr13:110741629-110742130</td>\n",
       "      <td>stim</td>\n",
       "      <td>-0.066564</td>\n",
       "      <td>0.034950</td>\n",
       "      <td>4.985318e-02</td>\n",
       "      <td>0.142614</td>\n",
       "      <td>0.091184</td>\n",
       "      <td>1.316118e-01</td>\n",
       "      <td>AMY_GABA</td>\n",
       "      <td>AMY</td>\n",
       "      <td>MW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120141</th>\n",
       "      <td>chr19:23980544-23981045</td>\n",
       "      <td>stim</td>\n",
       "      <td>-0.068154</td>\n",
       "      <td>0.035806</td>\n",
       "      <td>4.986629e-02</td>\n",
       "      <td>0.000641</td>\n",
       "      <td>0.086747</td>\n",
       "      <td>9.816067e-01</td>\n",
       "      <td>AMY_GABA</td>\n",
       "      <td>AMY</td>\n",
       "      <td>MW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120142</th>\n",
       "      <td>chr3:7366363-7366864</td>\n",
       "      <td>stim</td>\n",
       "      <td>-0.043569</td>\n",
       "      <td>0.022436</td>\n",
       "      <td>4.990000e-02</td>\n",
       "      <td>-0.036747</td>\n",
       "      <td>0.083953</td>\n",
       "      <td>6.686464e-01</td>\n",
       "      <td>AMY_GABA</td>\n",
       "      <td>AMY</td>\n",
       "      <td>MW</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>120143 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             gene    tx   de_coef     de_se        de_pval  \\\n",
       "0         chr15:75086569-75087070  stim  0.419844  0.017950  2.906623e-121   \n",
       "1          chr9:76305059-76305560  stim  0.682124  0.031764  6.342578e-102   \n",
       "2        chrX:143482812-143483313  stim  0.610629  0.028551  8.848753e-102   \n",
       "3       chr12:120028551-120029052  stim  0.397639  0.019140   3.674064e-96   \n",
       "4          chr2:98662045-98662546  stim  0.469970  0.024888   7.864128e-80   \n",
       "...                           ...   ...       ...       ...            ...   \n",
       "120138   chr1:167001277-167001778  stim -0.073560  0.037770   4.972503e-02   \n",
       "120139    chr18:88894092-88894593  stim -0.055283  0.028996   4.983414e-02   \n",
       "120140  chr13:110741629-110742130  stim -0.066564  0.034950   4.985318e-02   \n",
       "120141    chr19:23980544-23981045  stim -0.068154  0.035806   4.986629e-02   \n",
       "120142       chr3:7366363-7366864  stim -0.043569  0.022436   4.990000e-02   \n",
       "\n",
       "         dv_coef     dv_se       dv_pval    ctname region condition  \n",
       "0      -0.261757  0.092274  5.066288e-03  PFC_Glut    PFC        MC  \n",
       "1      -0.202397  0.088147  2.124100e-02  PFC_Glut    PFC        MC  \n",
       "2      -0.380780  0.125812  3.853716e-03  PFC_Glut    PFC        MC  \n",
       "3       0.007421  0.082149  9.290066e-01  PFC_Glut    PFC        MC  \n",
       "4      -0.503358  0.091436  7.121638e-08  PFC_Glut    PFC        MC  \n",
       "...          ...       ...           ...       ...    ...       ...  \n",
       "120138 -0.140123  0.147708  3.389143e-01  AMY_GABA    AMY        MW  \n",
       "120139 -0.000488  0.100180  9.568052e-01  AMY_GABA    AMY        MW  \n",
       "120140  0.142614  0.091184  1.316118e-01  AMY_GABA    AMY        MW  \n",
       "120141  0.000641  0.086747  9.816067e-01  AMY_GABA    AMY        MW  \n",
       "120142 -0.036747  0.083953  6.686464e-01  AMY_GABA    AMY        MW  \n",
       "\n",
       "[120143 rows x 11 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_DAR_ALL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>tx</th>\n",
       "      <th>de_coef</th>\n",
       "      <th>de_se</th>\n",
       "      <th>de_pval</th>\n",
       "      <th>dv_coef</th>\n",
       "      <th>dv_se</th>\n",
       "      <th>dv_pval</th>\n",
       "      <th>ctname</th>\n",
       "      <th>region</th>\n",
       "      <th>condition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr15:75086569-75087070</td>\n",
       "      <td>stim</td>\n",
       "      <td>0.419844</td>\n",
       "      <td>0.017950</td>\n",
       "      <td>2.906623e-121</td>\n",
       "      <td>-0.261757</td>\n",
       "      <td>0.092274</td>\n",
       "      <td>5.066288e-03</td>\n",
       "      <td>PFC_Glut</td>\n",
       "      <td>PFC</td>\n",
       "      <td>MC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr9:76305059-76305560</td>\n",
       "      <td>stim</td>\n",
       "      <td>0.682124</td>\n",
       "      <td>0.031764</td>\n",
       "      <td>6.342578e-102</td>\n",
       "      <td>-0.202397</td>\n",
       "      <td>0.088147</td>\n",
       "      <td>2.124100e-02</td>\n",
       "      <td>PFC_Glut</td>\n",
       "      <td>PFC</td>\n",
       "      <td>MC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX:143482812-143483313</td>\n",
       "      <td>stim</td>\n",
       "      <td>0.610629</td>\n",
       "      <td>0.028551</td>\n",
       "      <td>8.848753e-102</td>\n",
       "      <td>-0.380780</td>\n",
       "      <td>0.125812</td>\n",
       "      <td>3.853716e-03</td>\n",
       "      <td>PFC_Glut</td>\n",
       "      <td>PFC</td>\n",
       "      <td>MC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr12:120028551-120029052</td>\n",
       "      <td>stim</td>\n",
       "      <td>0.397639</td>\n",
       "      <td>0.019140</td>\n",
       "      <td>3.674064e-96</td>\n",
       "      <td>0.007421</td>\n",
       "      <td>0.082149</td>\n",
       "      <td>9.290066e-01</td>\n",
       "      <td>PFC_Glut</td>\n",
       "      <td>PFC</td>\n",
       "      <td>MC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr2:98662045-98662546</td>\n",
       "      <td>stim</td>\n",
       "      <td>0.469970</td>\n",
       "      <td>0.024888</td>\n",
       "      <td>7.864128e-80</td>\n",
       "      <td>-0.503358</td>\n",
       "      <td>0.091436</td>\n",
       "      <td>7.121638e-08</td>\n",
       "      <td>PFC_Glut</td>\n",
       "      <td>PFC</td>\n",
       "      <td>MC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120138</th>\n",
       "      <td>chr1:167001277-167001778</td>\n",
       "      <td>stim</td>\n",
       "      <td>-0.073560</td>\n",
       "      <td>0.037770</td>\n",
       "      <td>4.972503e-02</td>\n",
       "      <td>-0.140123</td>\n",
       "      <td>0.147708</td>\n",
       "      <td>3.389143e-01</td>\n",
       "      <td>AMY_GABA</td>\n",
       "      <td>AMY</td>\n",
       "      <td>MW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120139</th>\n",
       "      <td>chr18:88894092-88894593</td>\n",
       "      <td>stim</td>\n",
       "      <td>-0.055283</td>\n",
       "      <td>0.028996</td>\n",
       "      <td>4.983414e-02</td>\n",
       "      <td>-0.000488</td>\n",
       "      <td>0.100180</td>\n",
       "      <td>9.568052e-01</td>\n",
       "      <td>AMY_GABA</td>\n",
       "      <td>AMY</td>\n",
       "      <td>MW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120140</th>\n",
       "      <td>chr13:110741629-110742130</td>\n",
       "      <td>stim</td>\n",
       "      <td>-0.066564</td>\n",
       "      <td>0.034950</td>\n",
       "      <td>4.985318e-02</td>\n",
       "      <td>0.142614</td>\n",
       "      <td>0.091184</td>\n",
       "      <td>1.316118e-01</td>\n",
       "      <td>AMY_GABA</td>\n",
       "      <td>AMY</td>\n",
       "      <td>MW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120141</th>\n",
       "      <td>chr19:23980544-23981045</td>\n",
       "      <td>stim</td>\n",
       "      <td>-0.068154</td>\n",
       "      <td>0.035806</td>\n",
       "      <td>4.986629e-02</td>\n",
       "      <td>0.000641</td>\n",
       "      <td>0.086747</td>\n",
       "      <td>9.816067e-01</td>\n",
       "      <td>AMY_GABA</td>\n",
       "      <td>AMY</td>\n",
       "      <td>MW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120142</th>\n",
       "      <td>chr3:7366363-7366864</td>\n",
       "      <td>stim</td>\n",
       "      <td>-0.043569</td>\n",
       "      <td>0.022436</td>\n",
       "      <td>4.990000e-02</td>\n",
       "      <td>-0.036747</td>\n",
       "      <td>0.083953</td>\n",
       "      <td>6.686464e-01</td>\n",
       "      <td>AMY_GABA</td>\n",
       "      <td>AMY</td>\n",
       "      <td>MW</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>120143 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             gene    tx   de_coef     de_se        de_pval  \\\n",
       "0         chr15:75086569-75087070  stim  0.419844  0.017950  2.906623e-121   \n",
       "1          chr9:76305059-76305560  stim  0.682124  0.031764  6.342578e-102   \n",
       "2        chrX:143482812-143483313  stim  0.610629  0.028551  8.848753e-102   \n",
       "3       chr12:120028551-120029052  stim  0.397639  0.019140   3.674064e-96   \n",
       "4          chr2:98662045-98662546  stim  0.469970  0.024888   7.864128e-80   \n",
       "...                           ...   ...       ...       ...            ...   \n",
       "120138   chr1:167001277-167001778  stim -0.073560  0.037770   4.972503e-02   \n",
       "120139    chr18:88894092-88894593  stim -0.055283  0.028996   4.983414e-02   \n",
       "120140  chr13:110741629-110742130  stim -0.066564  0.034950   4.985318e-02   \n",
       "120141    chr19:23980544-23981045  stim -0.068154  0.035806   4.986629e-02   \n",
       "120142       chr3:7366363-7366864  stim -0.043569  0.022436   4.990000e-02   \n",
       "\n",
       "         dv_coef     dv_se       dv_pval    ctname region condition  \n",
       "0      -0.261757  0.092274  5.066288e-03  PFC_Glut    PFC        MC  \n",
       "1      -0.202397  0.088147  2.124100e-02  PFC_Glut    PFC        MC  \n",
       "2      -0.380780  0.125812  3.853716e-03  PFC_Glut    PFC        MC  \n",
       "3       0.007421  0.082149  9.290066e-01  PFC_Glut    PFC        MC  \n",
       "4      -0.503358  0.091436  7.121638e-08  PFC_Glut    PFC        MC  \n",
       "...          ...       ...           ...       ...    ...       ...  \n",
       "120138 -0.140123  0.147708  3.389143e-01  AMY_GABA    AMY        MW  \n",
       "120139 -0.000488  0.100180  9.568052e-01  AMY_GABA    AMY        MW  \n",
       "120140  0.142614  0.091184  1.316118e-01  AMY_GABA    AMY        MW  \n",
       "120141  0.000641  0.086747  9.816067e-01  AMY_GABA    AMY        MW  \n",
       "120142 -0.036747  0.083953  6.686464e-01  AMY_GABA    AMY        MW  \n",
       "\n",
       "[120143 rows x 11 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_DAR_ALL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_annotattion = utils.annotate_region(df_DAR_ALL,region_col='gene', bedfile='/data2st1/junyi/generegion_vM23/genebody_selected.bed')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>tx</th>\n",
       "      <th>de_coef</th>\n",
       "      <th>de_se</th>\n",
       "      <th>de_pval</th>\n",
       "      <th>dv_coef</th>\n",
       "      <th>dv_se</th>\n",
       "      <th>dv_pval</th>\n",
       "      <th>ctname</th>\n",
       "      <th>region</th>\n",
       "      <th>condition</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gstart</th>\n",
       "      <th>gend</th>\n",
       "      <th>strand</th>\n",
       "      <th>distance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr15:75086569-75087070</td>\n",
       "      <td>stim</td>\n",
       "      <td>0.419844</td>\n",
       "      <td>0.017950</td>\n",
       "      <td>2.906623e-121</td>\n",
       "      <td>-0.261757</td>\n",
       "      <td>0.092274</td>\n",
       "      <td>5.066288e-03</td>\n",
       "      <td>PFC_Glut</td>\n",
       "      <td>PFC</td>\n",
       "      <td>MC</td>\n",
       "      <td>Gm28502</td>\n",
       "      <td>ENSMUSG00000099452.1</td>\n",
       "      <td>75076209</td>\n",
       "      <td>75079094</td>\n",
       "      <td>-</td>\n",
       "      <td>-7476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr9:76305059-76305560</td>\n",
       "      <td>stim</td>\n",
       "      <td>0.682124</td>\n",
       "      <td>0.031764</td>\n",
       "      <td>6.342578e-102</td>\n",
       "      <td>-0.202397</td>\n",
       "      <td>0.088147</td>\n",
       "      <td>2.124100e-02</td>\n",
       "      <td>PFC_Glut</td>\n",
       "      <td>PFC</td>\n",
       "      <td>MC</td>\n",
       "      <td>Hcrtr2</td>\n",
       "      <td>ENSMUSG00000032360.16</td>\n",
       "      <td>76225879</td>\n",
       "      <td>76323856</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX:143482812-143483313</td>\n",
       "      <td>stim</td>\n",
       "      <td>0.610629</td>\n",
       "      <td>0.028551</td>\n",
       "      <td>8.848753e-102</td>\n",
       "      <td>-0.380780</td>\n",
       "      <td>0.125812</td>\n",
       "      <td>3.853716e-03</td>\n",
       "      <td>PFC_Glut</td>\n",
       "      <td>PFC</td>\n",
       "      <td>MC</td>\n",
       "      <td>Pak3</td>\n",
       "      <td>ENSMUSG00000031284.16</td>\n",
       "      <td>143518590</td>\n",
       "      <td>143797796</td>\n",
       "      <td>+</td>\n",
       "      <td>35278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr12:120028551-120029052</td>\n",
       "      <td>stim</td>\n",
       "      <td>0.397639</td>\n",
       "      <td>0.019140</td>\n",
       "      <td>3.674064e-96</td>\n",
       "      <td>0.007421</td>\n",
       "      <td>0.082149</td>\n",
       "      <td>9.290066e-01</td>\n",
       "      <td>PFC_Glut</td>\n",
       "      <td>PFC</td>\n",
       "      <td>MC</td>\n",
       "      <td>Tmem196</td>\n",
       "      <td>ENSMUSG00000048004.15</td>\n",
       "      <td>119945821</td>\n",
       "      <td>120021246</td>\n",
       "      <td>+</td>\n",
       "      <td>-7306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr2:98662045-98662546</td>\n",
       "      <td>stim</td>\n",
       "      <td>0.469970</td>\n",
       "      <td>0.024888</td>\n",
       "      <td>7.864128e-80</td>\n",
       "      <td>-0.503358</td>\n",
       "      <td>0.091436</td>\n",
       "      <td>7.121638e-08</td>\n",
       "      <td>PFC_Glut</td>\n",
       "      <td>PFC</td>\n",
       "      <td>MC</td>\n",
       "      <td>Gm10801</td>\n",
       "      <td>ENSMUSG00000075015.3</td>\n",
       "      <td>98662236</td>\n",
       "      <td>98664083</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120138</th>\n",
       "      <td>chr1:167001277-167001778</td>\n",
       "      <td>stim</td>\n",
       "      <td>-0.073560</td>\n",
       "      <td>0.037770</td>\n",
       "      <td>4.972503e-02</td>\n",
       "      <td>-0.140123</td>\n",
       "      <td>0.147708</td>\n",
       "      <td>3.389143e-01</td>\n",
       "      <td>AMY_GABA</td>\n",
       "      <td>AMY</td>\n",
       "      <td>MW</td>\n",
       "      <td>Fam78b</td>\n",
       "      <td>ENSMUSG00000060568.14</td>\n",
       "      <td>167001416</td>\n",
       "      <td>167091302</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120139</th>\n",
       "      <td>chr18:88894092-88894593</td>\n",
       "      <td>stim</td>\n",
       "      <td>-0.055283</td>\n",
       "      <td>0.028996</td>\n",
       "      <td>4.983414e-02</td>\n",
       "      <td>-0.000488</td>\n",
       "      <td>0.100180</td>\n",
       "      <td>9.568052e-01</td>\n",
       "      <td>AMY_GABA</td>\n",
       "      <td>AMY</td>\n",
       "      <td>MW</td>\n",
       "      <td>Socs6</td>\n",
       "      <td>ENSMUSG00000056153.15</td>\n",
       "      <td>88665223</td>\n",
       "      <td>88927481</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120140</th>\n",
       "      <td>chr13:110741629-110742130</td>\n",
       "      <td>stim</td>\n",
       "      <td>-0.066564</td>\n",
       "      <td>0.034950</td>\n",
       "      <td>4.985318e-02</td>\n",
       "      <td>0.142614</td>\n",
       "      <td>0.091184</td>\n",
       "      <td>1.316118e-01</td>\n",
       "      <td>AMY_GABA</td>\n",
       "      <td>AMY</td>\n",
       "      <td>MW</td>\n",
       "      <td>Gm33172</td>\n",
       "      <td>ENSMUSG00000114775.1</td>\n",
       "      <td>110683911</td>\n",
       "      <td>110686925</td>\n",
       "      <td>+</td>\n",
       "      <td>-54705</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120141</th>\n",
       "      <td>chr19:23980544-23981045</td>\n",
       "      <td>stim</td>\n",
       "      <td>-0.068154</td>\n",
       "      <td>0.035806</td>\n",
       "      <td>4.986629e-02</td>\n",
       "      <td>0.000641</td>\n",
       "      <td>0.086747</td>\n",
       "      <td>9.816067e-01</td>\n",
       "      <td>AMY_GABA</td>\n",
       "      <td>AMY</td>\n",
       "      <td>MW</td>\n",
       "      <td>Fam189a2</td>\n",
       "      <td>ENSMUSG00000071604.5</td>\n",
       "      <td>23972749</td>\n",
       "      <td>24031019</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120142</th>\n",
       "      <td>chr3:7366363-7366864</td>\n",
       "      <td>stim</td>\n",
       "      <td>-0.043569</td>\n",
       "      <td>0.022436</td>\n",
       "      <td>4.990000e-02</td>\n",
       "      <td>-0.036747</td>\n",
       "      <td>0.083953</td>\n",
       "      <td>6.686464e-01</td>\n",
       "      <td>AMY_GABA</td>\n",
       "      <td>AMY</td>\n",
       "      <td>MW</td>\n",
       "      <td>Pkia</td>\n",
       "      <td>ENSMUSG00000027499.12</td>\n",
       "      <td>7366668</td>\n",
       "      <td>7445366</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>120143 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             gene    tx   de_coef     de_se        de_pval  \\\n",
       "0         chr15:75086569-75087070  stim  0.419844  0.017950  2.906623e-121   \n",
       "1          chr9:76305059-76305560  stim  0.682124  0.031764  6.342578e-102   \n",
       "2        chrX:143482812-143483313  stim  0.610629  0.028551  8.848753e-102   \n",
       "3       chr12:120028551-120029052  stim  0.397639  0.019140   3.674064e-96   \n",
       "4          chr2:98662045-98662546  stim  0.469970  0.024888   7.864128e-80   \n",
       "...                           ...   ...       ...       ...            ...   \n",
       "120138   chr1:167001277-167001778  stim -0.073560  0.037770   4.972503e-02   \n",
       "120139    chr18:88894092-88894593  stim -0.055283  0.028996   4.983414e-02   \n",
       "120140  chr13:110741629-110742130  stim -0.066564  0.034950   4.985318e-02   \n",
       "120141    chr19:23980544-23981045  stim -0.068154  0.035806   4.986629e-02   \n",
       "120142       chr3:7366363-7366864  stim -0.043569  0.022436   4.990000e-02   \n",
       "\n",
       "         dv_coef     dv_se       dv_pval    ctname region condition gene_name  \\\n",
       "0      -0.261757  0.092274  5.066288e-03  PFC_Glut    PFC        MC   Gm28502   \n",
       "1      -0.202397  0.088147  2.124100e-02  PFC_Glut    PFC        MC    Hcrtr2   \n",
       "2      -0.380780  0.125812  3.853716e-03  PFC_Glut    PFC        MC      Pak3   \n",
       "3       0.007421  0.082149  9.290066e-01  PFC_Glut    PFC        MC   Tmem196   \n",
       "4      -0.503358  0.091436  7.121638e-08  PFC_Glut    PFC        MC   Gm10801   \n",
       "...          ...       ...           ...       ...    ...       ...       ...   \n",
       "120138 -0.140123  0.147708  3.389143e-01  AMY_GABA    AMY        MW    Fam78b   \n",
       "120139 -0.000488  0.100180  9.568052e-01  AMY_GABA    AMY        MW     Socs6   \n",
       "120140  0.142614  0.091184  1.316118e-01  AMY_GABA    AMY        MW   Gm33172   \n",
       "120141  0.000641  0.086747  9.816067e-01  AMY_GABA    AMY        MW  Fam189a2   \n",
       "120142 -0.036747  0.083953  6.686464e-01  AMY_GABA    AMY        MW      Pkia   \n",
       "\n",
       "                      gene_id     gstart       gend strand  distance  \n",
       "0        ENSMUSG00000099452.1   75076209   75079094      -     -7476  \n",
       "1       ENSMUSG00000032360.16   76225879   76323856      -         0  \n",
       "2       ENSMUSG00000031284.16  143518590  143797796      +     35278  \n",
       "3       ENSMUSG00000048004.15  119945821  120021246      +     -7306  \n",
       "4        ENSMUSG00000075015.3   98662236   98664083      +         0  \n",
       "...                       ...        ...        ...    ...       ...  \n",
       "120138  ENSMUSG00000060568.14  167001416  167091302      +         0  \n",
       "120139  ENSMUSG00000056153.15   88665223   88927481      -         0  \n",
       "120140   ENSMUSG00000114775.1  110683911  110686925      +    -54705  \n",
       "120141   ENSMUSG00000071604.5   23972749   24031019      -         0  \n",
       "120142  ENSMUSG00000027499.12    7366668    7445366      +         0  \n",
       "\n",
       "[120143 rows x 17 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_annotattion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_annotattion.to_csv('/data2st1/junyi/output/atac0416/dar/region_nt/df_annotattion_mememto.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_counts = df_DAR_ALL.groupby(['ctname','condition','region']).size().reset_index(name='counts')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ctname</th>\n",
       "      <th>condition</th>\n",
       "      <th>region</th>\n",
       "      <th>counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AMY_GABA</td>\n",
       "      <td>MC</td>\n",
       "      <td>AMY</td>\n",
       "      <td>8762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AMY_Glut</td>\n",
       "      <td>MC</td>\n",
       "      <td>AMY</td>\n",
       "      <td>8762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AMY_Glut</td>\n",
       "      <td>MW</td>\n",
       "      <td>AMY</td>\n",
       "      <td>3658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AMY_GABA</td>\n",
       "      <td>MW</td>\n",
       "      <td>AMY</td>\n",
       "      <td>3659</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HIP_GABA</td>\n",
       "      <td>MC</td>\n",
       "      <td>HIP</td>\n",
       "      <td>17485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HIP_Glut</td>\n",
       "      <td>MC</td>\n",
       "      <td>HIP</td>\n",
       "      <td>17485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HIP_GABA</td>\n",
       "      <td>MW</td>\n",
       "      <td>HIP</td>\n",
       "      <td>1636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HIP_Glut</td>\n",
       "      <td>MW</td>\n",
       "      <td>HIP</td>\n",
       "      <td>1636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>PFC_GABA</td>\n",
       "      <td>MC</td>\n",
       "      <td>PFC</td>\n",
       "      <td>500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>PFC_Glut</td>\n",
       "      <td>MC</td>\n",
       "      <td>PFC</td>\n",
       "      <td>500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>PFC_GABA</td>\n",
       "      <td>MW</td>\n",
       "      <td>PFC</td>\n",
       "      <td>28030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>PFC_Glut</td>\n",
       "      <td>MW</td>\n",
       "      <td>PFC</td>\n",
       "      <td>28030</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      ctname condition region  counts\n",
       "0   AMY_GABA        MC    AMY    8762\n",
       "2   AMY_Glut        MC    AMY    8762\n",
       "3   AMY_Glut        MW    AMY    3658\n",
       "1   AMY_GABA        MW    AMY    3659\n",
       "4   HIP_GABA        MC    HIP   17485\n",
       "6   HIP_Glut        MC    HIP   17485\n",
       "5   HIP_GABA        MW    HIP    1636\n",
       "7   HIP_Glut        MW    HIP    1636\n",
       "8   PFC_GABA        MC    PFC     500\n",
       "10  PFC_Glut        MC    PFC     500\n",
       "9   PFC_GABA        MW    PFC   28030\n",
       "11  PFC_Glut        MW    PFC   28030"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_counts.sort_values(by=['region','condition','counts'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.barplot(data=df_counts, y='ctname', x='counts', hue='condition', palette=['#FF7F0E','#1F77B4'], dodge=True,ci=None)\n",
    "plt.xlim (0, 1000)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.catplot(x=\"condition\",y='counts', hue=\"condition\",estimator='sum',ci=None, col=\"region\",order=['MW','MC'],hue_order=['MC','MW'], data=df_counts, kind=\"bar\", height=4, aspect=1.5,sharey=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for filename in files:\n",
    "    df_dar  = pd.read_csv(filename,index_col=0)\n",
    "    experimentname = re.split(r'[./]', filename)[-2]\n",
    "    experimentname\n",
    "    folder_name = os.path.dirname(filename)\n",
    "    df_dar_filtered = df_dar[(df_dar['pvals']<0.05) & (df_dar['logfoldchanges']>0) ]\n",
    "    df_dar_filtered.sort_values(by='logfoldchanges',ascending=False,inplace=True)\n",
    "    # For liftover\n",
    "\n",
    "    temp_df = df_dar_filtered.names.str.split(r'[ ,!\\-;:|]',expand=True)\n",
    "    temp_df['ID'] = temp_df.index\n",
    "    temp_df.to_csv(f\"{folder_name}/{experimentname}_DAR.bed\",header=False,index=False,sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def merge_region(region='PFC',condition=\"MC\",blacklist=['Doublet']):\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import subprocess\n",
    "# files = glob.glob('/data2st1/junyi/output/motif/ALL_*_wilcoxon_DAR.bed')\n",
    "# for filename in files:\n",
    "#     experimentname = re.split(r'[./]', filename)[-2]\n",
    "#     folder_name = os.path.dirname(filename)\n",
    "\n",
    "#     #!/home/junyichen/liftOver /data2st1/junyi/output/motif/PFC_Neuron_MW_wilcoxon_DAR.bed /data2st1/junyi/mm39ToMm10.over.chain.gz /data2st1/junyi/output/motif/PFC_Neuron_MW_wilcoxon_DAR_lifted.bed /data2st1/junyi/output/motif/PFC_Neuron_MW_wilcoxon_DAR_unmap.bed\n",
    "\n",
    "#     command_str = f\"/home/junyichen/liftOver {filename} /data2st1/junyi/mm39ToMm10.over.chain.gz {folder_name}/{experimentname}_lifted.bed {folder_name}/{experimentname}_unmap.bed\"\n",
    "#     print(command_str)\n",
    "#     subprocess.run(command_str, shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "files = glob.glob('/data2st1/junyi/output/motif/ALL_*_wilcoxon_DAR_lifted.bed')\n",
    "for filename in files:\n",
    "    destname = filename.replace(\"_lifted.bed\",\"_lifted_sorted.bed\")\n",
    "    #!sort -k1,1 -k2,2n {filename} > {destname}\n",
    "    command_str = f\"sort -k1,1 -k2,2n {filename} > {destname}\"\n",
    "    print(command_str)\n",
    "    subprocess.run(command_str, shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# !sort -k1,1 -k2,2n /data2st1/junyi/output/motif/HIP_Neuron_MC_wilcoxon_DAR_lifted.bed > /data2st1/junyi/output/motif/HIP_Neuron_MC_wilcoxon_DAR_lifted_sorted.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# !sort -k1,1 -k2,2n /data2st1/junyi/output/motif/PFC_Neuron_MC_wilcoxon_DAR_lifted.bed > /data2st1/junyi/output/motif/PFC_Neuron_MC_wilcoxon_DAR_lifted_sorted.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# !bedtools closest -a /data2st1/junyi/output/motif/PFC_Neuron_MC_wilcoxon_DAR_lifted_sorted.bed -b /data2st1/junyi/output/motif/genebody_selected_sorted.bed -D ref > /data2st1/junyi/output/motif/PFC_Neuron_MC_wilcoxon_DAR_gene.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# !bedtools closest -a /data2st1/junyi/output/motif/HIP_Neuron_MC_wilcoxon_DAR_lifted_sorted.bed -b /data2st1/junyi/output/motif/genebody_selected_sorted.bed -D ref > /data2st1/junyi/output/motif/HIP_Neuron_MC_wilcoxon_DAR_gene.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "files = glob.glob('/data2st1/junyi/output/motif/ALL_*_wilcoxon_DAR_lifted.bed')\n",
    "for filename in files:\n",
    "    command_str = ['pycistarget','cistarget','--cistarget_db_fname','/data2st1/junyi/scenic/mouse/motif/mm10_screen_v10_clust.regions_vs_motifs.rankings.feather'\n",
    "                   ,'--bed_fname',filename,'--species','mus_musculus','--auc_threshold','0.005','--nes_threshold','3.0','--rank_threshold','0.05'\n",
    "                   ,'--path_to_motif_annotations','/data2st1/junyi/scenic/mouse/motif/motifs-v10nr_clust-nr.mgi-m0.001-o0.0.tbl'\n",
    "                   ,'--output_folder','/data2st1/junyi/output/motif/','--write_html']\n",
    "    print(command_str)\n",
    "    subprocess.run(command_str)\n",
    "    \n",
    "# !pycistarget cistarget --cistarget_db_fname '/data2st1/junyi/scenic/mouse/motif/mm10_screen_v10_clust.regions_vs_motifs.rankings.feather' \\\n",
    "# --bed_fname '/data2st1/junyi/output/motif/AMY_Neuron_MC_wilcoxon_DAR_lifted.bed' \\\n",
    "# --species 'mus_musculus' \\\n",
    "# --auc_threshold 0.005 \\\n",
    "# --nes_threshold 3.0 \\\n",
    "# --rank_threshold 0.05 \\\n",
    "# --path_to_motif_annotations '/data2st1/junyi/scenic/mouse/motif/motifs-v10nr_clust-nr.mgi-m0.001-o0.0.tbl' \\\n",
    "# --output_folder '/data2st1/junyi/output/motif/' \\\n",
    "# --write_html "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import h5py\n",
    "dict_table = {}\n",
    "cell_types = ['Neuron','Astro-Epen','Vascular','Immune','OPC-Oligo']\n",
    "for celltype in cell_types:\n",
    "    with h5py.File(f'/data2st1/junyi/output/motif/motif_enrichment_cistarget_ALL_{celltype}_wilcoxon_DAR_lifted.hdf5', 'r') as f:\n",
    "        # Open the HDF5 file\n",
    "        # List all groups and datasets in the file\n",
    "        print(\"Keys in the file:\", list(f.keys()))\n",
    "        \n",
    "\n",
    "        # # Access a specific dataset\n",
    "        expname = f'ALL_{celltype}_wilcoxon_DAR_lifted'\n",
    "        dataset = f[expname]  # Replace with your dataset name\n",
    "        # print(\"Shape of the dataset:\", dataset.shape)\n",
    "        # print(\"Data type of the dataset:\", dataset.dtype)\n",
    "        dict_dataset = {}\n",
    "        #region = f[expname]['motif_hits']['region_set']['metacluster_33.8'][0:10]\n",
    "\n",
    "        table = f[expname]['motif_enrichment']['table'][:]\n",
    "        for key in dataset.keys():\n",
    "            dict_dataset[key] = dataset[key]\n",
    "        # print(\"Data:\", data)\n",
    "        dict_table[expname] = table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for keys in dict_table.keys():\n",
    "    print(keys)\n",
    "    print(dict_table[keys][:5])\n",
    "    #Astro : id, orthology, direcrt, id, msimilarity,logo\n",
    "    #Neuron : id,logo,orthology, id, direct, msimilarity\n",
    "    #Vascular : id,logo,orthology,msimilarity,id, direct"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "list_table = []\n",
    "list_key = []\n",
    "list_TFs = []\n",
    "\n",
    "df_TF_celltype = pd.DataFrame()\n",
    "\n",
    "for key in dict_table.keys():\n",
    "    data = dict_table[key]\n",
    "    for row in data:\n",
    "        list_table.append(row[0].decode('utf-8'))\n",
    "        list_key.append(key)\n",
    "        TFs_tmp=[]\n",
    "        for i in range(1, len(row)-2):\n",
    "            if 'img' in row[i][0].decode('utf-8') or 'wilcoxon' in row[i][0].decode('utf-8'):\n",
    "                continue\n",
    "            TFs_tmp+=([ tf.decode('utf-8') for tf in row[i]])\n",
    "\n",
    "        df_tfs = pd.DataFrame({'TF':\",\".join(TFs_tmp).split(',')})\n",
    "        df_tfs['NES'] = row[-2][0]\n",
    "        df_tfs['AUC'] = row[-2][1]\n",
    "        df_tfs['Rank'] = row[-2][2]\n",
    "        df_tfs['celltype.L1'] = key.split('_')[1]\n",
    "        df_tfs['id'] = row[0].decode('utf-8')\n",
    "\n",
    "        list_TFs.append(set(TFs_tmp))\n",
    "        df_TF_celltype = pd.concat([df_TF_celltype,df_tfs],axis=0)\n",
    "        \n",
    "df_TF = pd.DataFrame({'TF':list_table,'key':list_key,'TFs':list_TFs})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_TF_celltype.drop_duplicates(inplace=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_TF_celltype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_RSS = pd.read_csv(\"/home/junyichen/code/scmmd/data/RSS_Score(only pos).csv\",index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_RSS['TF']=df_RSS['Topic'].str.replace('[()+]','')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_to_category = {\n",
    "    # Astro-Epen\n",
    "    'Astro-Gfaphigh': 'Astro-Epen',\n",
    "    'Astro-TE': 'Astro-Epen',\n",
    "    \n",
    "    # Immune\n",
    "    'Microglia': 'Immune',\n",
    "    'PVM': 'Immune',\n",
    "    \n",
    "    # Neuron\n",
    "    'L2/3 IT': 'Neuron',\n",
    "    'L4/5 IT': 'Neuron',\n",
    "    'L5 ET': 'Neuron',\n",
    "    'L6 CT': 'Neuron',\n",
    "    'L6 IT': 'Neuron',\n",
    "    'L6b': 'Neuron',\n",
    "    'Lamp5 GABA': 'Neuron',\n",
    "    'Pvalb GABA': 'Neuron',\n",
    "    'Pvalb Vipr2 GABA': 'Neuron',\n",
    "    'Sst GABA': 'Neuron',\n",
    "    'Vip GABA': 'Neuron',\n",
    "    \n",
    "    # OPC-Oligo\n",
    "    'COP': 'OPC-Oligo',\n",
    "    'MFOL': 'OPC-Oligo',\n",
    "    'MOL': 'OPC-Oligo',\n",
    "    'NFOL': 'OPC-Oligo',\n",
    "    'NP': 'OPC-Oligo',\n",
    "    'OPC': 'OPC-Oligo',\n",
    "    \n",
    "    # Vascular\n",
    "    'Arachnoid barrier cell': 'Vascular',\n",
    "    'Car3': 'Vascular',\n",
    "    'Endothelial cell': 'Vascular',\n",
    "    'Pericyte': 'Vascular',\n",
    "    'VLMC': 'Vascular'\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_RSS['celltype.L1'] = df_RSS['celltype.L2'].map(cell_to_category)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_TF_celltype.groupby(['celltype.L1']).nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_RSS.groupby(['celltype.L1']).nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_merged = df_RSS.merge(df_TF_celltype,left_on=['TF','celltype.L1'],right_on=['TF','celltype.L1'],how='inner')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_merged.sort_values(by='RSS',ascending=False,inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_merged.drop_duplicates(set(['TF','celltype.L1']),keep='first').to_csv('/data2st1/junyi/output/motif/TF_RSS.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_merged.groupby('celltype.L1').nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ct = pd.crosstab(df_merged['celltype'],df_merged['celltype.L1'],values=df_merged['RSS'],aggfunc='count')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ct"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "set(df_merged['celltype']\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "files = glob.glob('/data2st1/junyi/output/motif/ALL_*_wilcoxon_DAR_lifted_sorted.bed')\n",
    "for filename in files:\n",
    "    destname = filename.replace(\"_lifted_sorted.bed\",\"_lifted_sorted_gene.bed\")\n",
    "    command_str = f\"bedtools closest -a {filename} -b /data2st1/junyi/output/motif/genebody_selected_sorted.bed -D ref > {destname}\"\n",
    "    print(command_str)\n",
    "    subprocess.run(command_str, shell=True)\n",
    "# !bedtools closest -a /data2st1/junyi/output/motif/PFC_Neuron_MC_wilcoxon_DAR_lifted_sorted.bed -b /data2st1/junyi/output/motif/genebody_selected_sorted.bed -D ref > /data2st1/junyi/output/motif/PFC_Neuron_MC_wilcoxon_DAR_gene.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!pycistarget cistarget --cistarget_db_fname '/data2st1/junyi/scenic/mouse/motif/mm10_screen_v10_clust.regions_vs_motifs.rankings.feather' \\\n",
    "--bed_fname '/data2st1/junyi/output/motif/AMY_Neuron_MC_wilcoxon_DAR_lifted.bed' \\\n",
    "--species 'mus_musculus' \\\n",
    "--auc_threshold 0.005 \\\n",
    "--nes_threshold 3.0 \\\n",
    "--rank_threshold 0.05 \\\n",
    "--path_to_motif_annotations '/data2st1/junyi/scenic/mouse/motif/motifs-v10nr_clust-nr.mgi-m0.001-o0.0.tbl' \\\n",
    "--output_folder '/data2st1/junyi/output/motif/' \\\n",
    "--write_html "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "files = glob.glob('/data2st1/junyi/output/motif/*MW*wilcoxon.csv')\n",
    "for filename in files:\n",
    "    df_dar  = pd.read_csv(filename,index_col=0)\n",
    "    experimentname = re.split(r'[./]', filename)[-2]\n",
    "    experimentname\n",
    "    folder_name = os.path.dirname(filename)\n",
    "    df_dar_filtered = df_dar[(df_dar['pvals']<0.05) & (df_dar['logfoldchanges']>0) ]\n",
    "    df_dar_filtered.sort_values(by='logfoldchanges',ascending=False,inplace=True)\n",
    "    # For liftover\n",
    "    df_dar_filtered.names.str.split(r'[ ,!\\-;:|]',expand=True).to_csv(f\"{folder_name}/{experimentname}_DAR.bed\",header=False,index=False,sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "files = glob.glob('/data2st1/junyi/output/motif/*MW*wilcoxon_DAR.bed')\n",
    "for filename in files:\n",
    "    experimentname = re.split(r'[./]', filename)[-2]\n",
    "    folder_name = os.path.dirname(filename)\n",
    "\n",
    "    #!/home/junyichen/liftOver /data2st1/junyi/output/motif/PFC_Neuron_MW_wilcoxon_DAR.bed /data2st1/junyi/mm39ToMm10.over.chain.gz /data2st1/junyi/output/motif/PFC_Neuron_MW_wilcoxon_DAR_lifted.bed /data2st1/junyi/output/motif/PFC_Neuron_MW_wilcoxon_DAR_unmap.bed\n",
    "\n",
    "    command_str = f\"/home/junyichen/liftOver {filename} /data2st1/junyi/mm39ToMm10.over.chain.gz {folder_name}/{experimentname}_lifted.bed {folder_name}/{experimentname}_unmap.bed\"\n",
    "    print(command_str)\n",
    "    subprocess.run(command_str, shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "files = glob.glob('/data2st1/junyi/output/motif/*MW*wilcoxon_DAR_lifted.bed')\n",
    "for filename in files:\n",
    "    destname = filename.replace(\"_lifted.bed\",\"_lifted_sorted.bed\")\n",
    "    #!sort -k1,1 -k2,2n {filename} > {destname}\n",
    "    command_str = f\"sort -k1,1 -k2,2n {filename} > {destname}\"\n",
    "    print(command_str)\n",
    "    subprocess.run(command_str, shell=True)\n",
    "files = glob.glob('/data2st1/junyi/output/motif/*MW*wilcoxon_DAR_lifted_sorted.bed')\n",
    "for filename in files:\n",
    "    destname = filename.replace(\"_lifted_sorted.bed\",\"_lifted_sorted_gene.bed\")\n",
    "    command_str = f\"bedtools closest -a {filename} -b /data2st1/junyi/output/motif/genebody_selected_sorted.bed -D ref > {destname}\"\n",
    "    print(command_str)\n",
    "    subprocess.run(command_str, shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from pyliftover import LiftOver\n",
    "# lo = LiftOver('/data2st1/junyi/mm10ToMm39.over.chain.gz')\n",
    "# lo.convert_coordinate('chr1', 1000000, '-')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_frame = pd.DataFrame()\n",
    "files = glob.glob('/data2st1/junyi/output/motif/*MW*wilcoxon_DAR_lifted_sorted_gene.bed')\n",
    "for filename in files:\n",
    "    experimentname = re.split(r'[./]', filename)[-2]\n",
    "    region = re.split(r'_',experimentname)[0]\n",
    "    celltype = re.split(r'_',experimentname)[1]\n",
    "    condition = re.split(r'_',experimentname)[2]\n",
    "    try:\n",
    "        df = pd.read_csv(filename,sep='\\t',header=None)\n",
    "    except:\n",
    "        continue\n",
    "    df.columns = ['chrom','start','end','chrom2','start2','end2','score','starnd','gene_name','gene_id','annotation','distance']\n",
    "    df['region'] = region\n",
    "    df['celltype'] = celltype\n",
    "    df['condition'] = condition\n",
    "    df['name'] = df['chrom'] + ':' + df['start'].astype(str) + '-' + df['end'].astype(str)\n",
    "    # df_score = pd.read_csv(f\"/data2st1/junyi/output/motif/{region}_{celltype}_{condition}_wilcoxon.csv\",index_col=0)\n",
    "    # df['logfoldchanges'] = df_score.set_index('names').loc[df['name'],'logfoldchanges'].values\n",
    "    # df['pvals'] = df_score.set_index('names').loc[df['name'],'pvals'].values\n",
    "    result_frame = pd.concat([result_frame,df])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ax = sns.countplot(data=result_frame, x='celltype', hue='region', dodge=False, palette='Set2',)\n",
    "ax.set_ylim(0,6000)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "keep1=result_frame.sort_values(by='starnd',ascending=True).drop_duplicates(subset=['chrom','start','end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "keep1['generegion'] = 'genebody'\n",
    "keep1.loc[keep1['distance']<0,'generegion'] = 'downstream'\n",
    "promoter = keep1.loc[(keep1['distance']>0) & (keep1['distance']<=2000) & (keep1['starnd']=='+')].index\n",
    "distal = keep1.loc[(keep1['distance']>0) & (keep1['distance']>=2000) & (keep1['starnd']=='+')].index\n",
    "keep1.loc[promoter,'generegion'] = 'promoter'\n",
    "keep1.loc[distal,'generegion'] = 'distal'\n",
    "\n",
    "promoter = keep1.loc[(keep1['distance']<0) & (keep1['distance']>=-2000) & (keep1['starnd']=='-')].index\n",
    "distal = keep1.loc[(keep1['distance']<0) & (keep1['distance']<=-2000) & (keep1['starnd']=='-')].index\n",
    "keep1.loc[promoter,'generegion'] = 'promoter'\n",
    "keep1.loc[distal,'generegion'] = 'distal'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "keep1=keep1.loc[keep1['start2']>0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "keep1.sort_values(by='starnd',ascending=True).drop_duplicates(subset=['chrom','start','end'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ax = sns.countplot(data=keep1, x='celltype', hue='generegion', dodge=False, palette='Set2',)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.read_csv('/data2st1/junyi/output/motif/AMY_Neuron_MC_wilcoxon.csv',index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert hg38 to hg19\n",
    "# !/home/junyichen/liftOver /data2st1/junyi/output/motif/HIP_Neuron_MW_wilcoxon_DAR.bed /data2st1/junyi/mm39ToMm10.over.chain.gz /data2st1/junyi/output/motif/HIP_Neuron_MW_wilcoxon_DAR_lifted.bed /data2st1/junyi/output/motif/HIP_Neuron_MW_wilcoxon_DAR_unmap.bed\n",
    "# !/home/junyichen/liftOver /data2st1/junyi/output/motif/PFC_Neuron_MW_wilcoxon_DAR.bed /data2st1/junyi/mm39ToMm10.over.chain.gz /data2st1/junyi/output/motif/PFC_Neuron_MW_wilcoxon_DAR_lifted.bed /data2st1/junyi/output/motif/PFC_Neuron_MW_wilcoxon_DAR_unmap.bed\n",
    "# !/home/junyichen/liftOver /data2st1/junyi/output/motif/AMY_Neuron_MW_wilcoxon_DAR.bed /data2st1/junyi/mm39ToMm10.over.chain.gz /data2st1/junyi/output/motif/AMY_Neuron_MW_wilcoxon_DAR_lifted.bed /data2st1/junyi/output/motif/AMY_Neuron_MW_wilcoxon_DAR_unmap.bed\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!pycistarget cistarget --cistarget_db_fname '/data2st1/junyi/scenic/mouse/motif/mm10_screen_v10_clust.regions_vs_motifs.rankings.feather' \\\n",
    "--bed_fname '/data2st1/junyi/output/motif/HIP_Neuron_MW_wilcoxon_DAR_lifted.bed' \\\n",
    "--species 'mus_musculus' \\\n",
    "--auc_threshold 0.005 \\\n",
    "--nes_threshold 3.0 \\\n",
    "--rank_threshold 0.05 \\\n",
    "--path_to_motif_annotations '/data2st1/junyi/scenic/mouse/motif/motifs-v10nr_clust-nr.mgi-m0.001-o0.0.tbl' \\\n",
    "--output_folder '/data2st1/junyi/output/motif/' \\\n",
    "--write_html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!pycistarget cistarget --cistarget_db_fname '/data2st1/junyi/scenic/mouse/motif/mm10_screen_v10_clust.regions_vs_motifs.rankings.feather' \\\n",
    "--bed_fname '/data2st1/junyi/output/motif/AMY_Neuron_MW_wilcoxon_DAR_lifted.bed' \\\n",
    "--species 'mus_musculus' \\\n",
    "--auc_threshold 0.005 \\\n",
    "--nes_threshold 3.0 \\\n",
    "--rank_threshold 0.05 \\\n",
    "--path_to_motif_annotations '/data2st1/junyi/scenic/mouse/motif/motifs-v10nr_clust-nr.mgi-m0.001-o0.0.tbl' \\\n",
    "--output_folder '/data2st1/junyi/output/motif/' \\\n",
    "--write_html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!pycistarget cistarget --cistarget_db_fname '/data2st1/junyi/scenic/mouse/motif/mm10_screen_v10_clust.regions_vs_motifs.rankings.feather' \\\n",
    "--bed_fname '/data2st1/junyi/output/motif/PFC_Neuron_MW_wilcoxon_DAR_lifted.bed' \\\n",
    "--species 'mus_musculus' \\\n",
    "--auc_threshold 0.005 \\\n",
    "--nes_threshold 3.0 \\\n",
    "--rank_threshold 0.05 \\\n",
    "--path_to_motif_annotations '/data2st1/junyi/scenic/mouse/motif/motifs-v10nr_clust-nr.mgi-m0.001-o0.0.tbl' \\\n",
    "--output_folder '/data2st1/junyi/output/motif/' \\\n",
    "--write_html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import h5py\n",
    "import copy\n",
    "\n",
    "regions = ['AMY','HIP','PFC']\n",
    "celltypes = ['Neuron']\n",
    "\n",
    "dict_table = {}\n",
    "\n",
    "for region in regions:\n",
    "    for celltype in celltypes:\n",
    "        with h5py.File(f'/data2st1/junyi/output/motif/motif_enrichment_cistarget_{region}_{celltype}_MC_wilcoxon_DAR_lifted.hdf5', 'r') as f:\n",
    "            # Open the HDF5 file\n",
    "            # List all groups and datasets in the file\n",
    "            print(\"Keys in the file:\", list(f.keys()))\n",
    "            \n",
    "\n",
    "            # # Access a specific dataset\n",
    "            expname = f'{region}_{celltype}_MC_wilcoxon_DAR_lifted'\n",
    "            dataset = f[expname]  # Replace with your dataset name\n",
    "            # print(\"Shape of the dataset:\", dataset.shape)\n",
    "            # print(\"Data type of the dataset:\", dataset.dtype)\n",
    "\n",
    "            dict_dataset = {}\n",
    "\n",
    "            #region = f[expname]['motif_hits']['region_set']['metacluster_33.8'][0:10]\n",
    "\n",
    "            table = f[expname]['motif_enrichment']['table'][:]\n",
    "            for key in dataset.keys():\n",
    "                dict_dataset[key] = dataset[key]\n",
    "            # print(\"Data:\", data)\n",
    "            dict_table[expname] = table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in dict_table['AMY_Neuron_MC_wilcoxon_DAR_lifted']:\n",
    "    print(i[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "region_table = {}\n",
    "\n",
    "for region in regions:\n",
    "    for celltype in celltypes:\n",
    "        with h5py.File(f'/data2st1/junyi/output/motif/motif_enrichment_cistarget_{region}_{celltype}_MC_wilcoxon_DAR_lifted.hdf5', 'r') as f:\n",
    "            # Open the HDF5 file\n",
    "            # List all groups and datasets in the file\n",
    "            print(\"Keys in the file:\", list(f.keys()))\n",
    "            \n",
    "\n",
    "            # # Access a specific dataset\n",
    "            expname = f'{region}_{celltype}_MC_wilcoxon_DAR_lifted'\n",
    "            dataset = f[expname]  # Replace with your dataset name\n",
    "            # print(\"Shape of the dataset:\", dataset.shape)\n",
    "            # print(\"Data type of the dataset:\", dataset.dtype)\n",
    "\n",
    "            dict_dataset = {}\n",
    "\n",
    "            region = f[expname]['motif_hits']['region_set']['metacluster_33.8'][:]\n",
    "            region_table[expname] = region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "list_table = []\n",
    "list_key = []\n",
    "\n",
    "for key in dict_table.keys():\n",
    "    data = dict_table[key]\n",
    "    for row in data:\n",
    "        list_table.append(row[0])\n",
    "        list_key.append(key)\n",
    "df_TF = pd.DataFrame({'TF':list_table,'key':list_key})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_TF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_count = df_TF.groupby('TF').count().sort_values(by='key',ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_count[df_count['key']>2].index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_regions = pd.DataFrame()\n",
    "\n",
    "for region in regions:\n",
    "    for celltype in celltypes:\n",
    "        with h5py.File(f'/data2st1/junyi/output/motif/motif_enrichment_cistarget_{region}_{celltype}_MC_wilcoxon_DAR_lifted.hdf5', 'r') as f:\n",
    "            # Open the HDF5 file\n",
    "            # List all groups and datasets in the file\n",
    "            print(\"Keys in the file:\", list(f.keys()))\n",
    "            \n",
    "\n",
    "            # # Access a specific dataset\n",
    "            expname = f'{region}_{celltype}_MC_wilcoxon_DAR_lifted'\n",
    "            dataset = f[expname]  # Replace with your dataset name\n",
    "            # print(\"Shape of the dataset:\", dataset.shape)\n",
    "            # print(\"Data type of the dataset:\", dataset.dtype)\n",
    "\n",
    "\n",
    "            for TF in df_count[df_count['key']>2].index:\n",
    "                region = f[expname]['motif_hits']['region_set'][TF][:]\n",
    "                decoded_data = [item.decode('utf-8') for item in region]\n",
    "                df_region = pd.DataFrame(decoded_data,columns=['region'])\n",
    "                df_region['expreiment'] = expname\n",
    "                df_region['TF'] = TF.decode('utf-8')\n",
    "                if len(df_regions)>0:\n",
    "                    df_regions = pd.concat([df_regions,df_region],axis=0,ignore_index=True)\n",
    "                else:\n",
    "                    df_regions = df_region\n",
    "            #region = f[expname]['motif_hits']['region_set']['metacluster_33.8'][:]\n",
    "            #region_table[expname] = df_regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_count[df_count['key']>2].index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_regions.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "region_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.countplot(data=df_TF, x='key', dodge=False, palette='Set2',)\n",
    "plt.xticks(rotation=60)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for key in dict_table.keys():\n",
    "    print(len(dict_table[key]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "table"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "snapatac2",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
