{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%\n",
    "import pandas as pd\n",
    "import tarfile\n",
    "import io\n",
    "from ALLCools.mcds import MCDS\n",
    "import pyBigWig\n",
    "import pandas as pd# %%\n",
    "from wmb import cemba, mm10\n",
    "import glob as glob\n",
    "import numpy as np\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import pybedtools\n",
    "\n",
    "# %%\n",
    "black_list_path = mm10.ENCODE_BLACKLIST_PATH\n",
    "\n",
    "\n",
    "# %%\n",
    "black_list_path\n",
    "\n",
    "# %%\n",
    "mch_pattern = 'CHN'\n",
    "mcg_pattern = 'CGN'\n",
    "n_top_feature = 20000\n",
    "\n",
    "# %%\n",
    "gene_meta_path = '/home/junyichen/code/whole_mouse_brain/wmb/files/modified_gencode.vM23.primary_assembly.annotation.gene.flat.tsv.gz'\n",
    "\n",
    "# %%\n",
    "chrom_to_remove = ['chrM', 'chrX', 'chrY']\n",
    "\n",
    "\n",
    "mcds_expression = MCDS.open(\n",
    "    '/data2st1/junyi/methlyatlas/mCseq/data.nemoarchive.org/biccn/grant/u19_cemba/ecker/epigenome/cellgroup/mCseq3/mouse/processed/counts/CEMBA.snmC.L4Region.AIBS_TENX.log1pCPM.zarr/',\n",
    "     obs_dim='CellGroup'\n",
    ")\n",
    "\n",
    "\n",
    "# %%\n",
    "var_dim = \"geneslop2k-vm23\"\n",
    "mcds = MCDS.open('/data2st1/junyi/methlyatlas/mCseq/data.nemoarchive.org/biccn/grant/u19_cemba/ecker/epigenome/cellgroup/mCseq3/mouse/processed/counts/CEMBA.snmC.L4RegionAgg.zarr',\n",
    "                   obs_dim='L4Region',var_dim=var_dim)\n",
    "\n",
    "\n",
    "# %%\n",
    "abs(mcds['geneslop2k-vm23_start']-mcds['geneslop2k-vm23_end']).min()\n",
    "\n",
    "# %%\n",
    "use_gene_meta = pd.read_csv(\"/home/junyichen/code/scmmd/output/GeneMetadata.csv.gz\",index_col=0)\n",
    "\n",
    "# %%# %%\n",
    "df_region = pd.read_csv('/data2st1/junyi/methlyatlas/atac/Subclass.names.map.csv.gz',index_col=0)\n",
    "\n",
    "# %%\n",
    "df_region\n",
    "\n",
    "# %%\n",
    "df_meta_integrateion = pd.read_csv(\"/data2st1/junyi/methlyatlas/mCseq/CEMBA.mC.Metadata/CEMBA.mC.Metadata.csv\")\n",
    "\n",
    "# %%\n",
    "df_meta_integrateion.drop_duplicates(subset='CellGroup',inplace=True)\n",
    "\n",
    "# %%\n",
    "df_meta_integrateion.set_index('CellGroup',inplace=True)\n",
    "\n",
    "# %%\n",
    "df_meta_integrateion.loc[mcds[\"L4Region\"]]\n",
    "\n",
    "# %%\n",
    "df_meta_integrateion\n",
    "\n",
    "\n",
    "bw_files_mseq = glob.glob(\"/data2st1/junyi/methlyatlas/mCseq/data.nemoarchive.org/biccn/grant/u19_cemba/ecker/epigenome/cellgroup/mCseq3/mouse/processed/other/*CGN*bw\")\n",
    "\n",
    "cell_groups = []\n",
    "methtypes =  []\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyBigWig\n",
    "bb_file = pyBigWig.open(\"/data2st1/junyi/generegion_vM33/encodeCcreCombined.bb\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_status</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>level</th>\n",
       "      <th>gene_version</th>\n",
       "      <th>mgi_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>tag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000051951</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3205901</td>\n",
       "      <td>3671498</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>MGI:3528744</td>\n",
       "      <td>OTTMUSG00000026353.2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000102343</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3905739</td>\n",
       "      <td>3986215</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:5610609</td>\n",
       "      <td>OTTMUSG00000049934.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000025900</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3999557</td>\n",
       "      <td>4409241</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>12</td>\n",
       "      <td>MGI:1341105</td>\n",
       "      <td>OTTMUSG00000049985.3</td>\n",
       "      <td>overlapping_locus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000025902</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4490931</td>\n",
       "      <td>4497354</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>13</td>\n",
       "      <td>MGI:107543</td>\n",
       "      <td>OTTMUSG00000050014.7</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000104238</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4496551</td>\n",
       "      <td>4499558</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:5610815</td>\n",
       "      <td>OTTMUSG00000050024.2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000100574</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>89636206</td>\n",
       "      <td>89639438</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:5579380</td>\n",
       "      <td>OTTMUSG00000047315.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000100608</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>89713424</td>\n",
       "      <td>89745532</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:5440224</td>\n",
       "      <td>OTTMUSG00000047316.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000102011</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>89927120</td>\n",
       "      <td>89935013</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:5579636</td>\n",
       "      <td>OTTMUSG00000047349.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000096178</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>90401248</td>\n",
       "      <td>90433263</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>MGI:5434193</td>\n",
       "      <td>OTTMUSG00000047352.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000096768</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>90784738</td>\n",
       "      <td>90816465</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>MGI:6096131</td>\n",
       "      <td>OTTMUSG00000074820.2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>27651 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   chrom  source feature     start       end score strand  \\\n",
       "ENSMUSG00000051951  chr1  HAVANA    gene   3205901   3671498     .      -   \n",
       "ENSMUSG00000102343  chr1  HAVANA    gene   3905739   3986215     .      -   \n",
       "ENSMUSG00000025900  chr1  HAVANA    gene   3999557   4409241     .      -   \n",
       "ENSMUSG00000025902  chr1  HAVANA    gene   4490931   4497354     .      -   \n",
       "ENSMUSG00000104238  chr1  HAVANA    gene   4496551   4499558     .      +   \n",
       "...                  ...     ...     ...       ...       ...   ...    ...   \n",
       "ENSMUSG00000100574  chrY  HAVANA    gene  89636206  89639438     .      -   \n",
       "ENSMUSG00000100608  chrY  HAVANA    gene  89713424  89745532     .      -   \n",
       "ENSMUSG00000102011  chrY  HAVANA    gene  89927120  89935013     .      -   \n",
       "ENSMUSG00000096178  chrY  HAVANA    gene  90401248  90433263     .      +   \n",
       "ENSMUSG00000096768  chrY  HAVANA    gene  90784738  90816465     .      +   \n",
       "\n",
       "                   phase  transcript_id       gene_type  ...  transcript_type  \\\n",
       "ENSMUSG00000051951     .            NaN  protein_coding  ...              NaN   \n",
       "ENSMUSG00000102343     .            NaN          lncRNA  ...              NaN   \n",
       "ENSMUSG00000025900     .            NaN  protein_coding  ...              NaN   \n",
       "ENSMUSG00000025902     .            NaN  protein_coding  ...              NaN   \n",
       "ENSMUSG00000104238     .            NaN          lncRNA  ...              NaN   \n",
       "...                  ...            ...             ...  ...              ...   \n",
       "ENSMUSG00000100574     .            NaN          lncRNA  ...              NaN   \n",
       "ENSMUSG00000100608     .            NaN  protein_coding  ...              NaN   \n",
       "ENSMUSG00000102011     .            NaN          lncRNA  ...              NaN   \n",
       "ENSMUSG00000096178     .            NaN  protein_coding  ...              NaN   \n",
       "ENSMUSG00000096768     .            NaN          lncRNA  ...              NaN   \n",
       "\n",
       "                   transcript_status  transcript_name  exon_number  exon_id  \\\n",
       "ENSMUSG00000051951               NaN              NaN          NaN      NaN   \n",
       "ENSMUSG00000102343               NaN              NaN          NaN      NaN   \n",
       "ENSMUSG00000025900               NaN              NaN          NaN      NaN   \n",
       "ENSMUSG00000025902               NaN              NaN          NaN      NaN   \n",
       "ENSMUSG00000104238               NaN              NaN          NaN      NaN   \n",
       "...                              ...              ...          ...      ...   \n",
       "ENSMUSG00000100574               NaN              NaN          NaN      NaN   \n",
       "ENSMUSG00000100608               NaN              NaN          NaN      NaN   \n",
       "ENSMUSG00000102011               NaN              NaN          NaN      NaN   \n",
       "ENSMUSG00000096178               NaN              NaN          NaN      NaN   \n",
       "ENSMUSG00000096768               NaN              NaN          NaN      NaN   \n",
       "\n",
       "                    level  gene_version       mgi_id           havana_gene  \\\n",
       "ENSMUSG00000051951      2             5  MGI:3528744  OTTMUSG00000026353.2   \n",
       "ENSMUSG00000102343      2             1  MGI:5610609  OTTMUSG00000049934.1   \n",
       "ENSMUSG00000025900      2            12  MGI:1341105  OTTMUSG00000049985.3   \n",
       "ENSMUSG00000025902      2            13   MGI:107543  OTTMUSG00000050014.7   \n",
       "ENSMUSG00000104238      2             1  MGI:5610815  OTTMUSG00000050024.2   \n",
       "...                   ...           ...          ...                   ...   \n",
       "ENSMUSG00000100574      2             1  MGI:5579380  OTTMUSG00000047315.1   \n",
       "ENSMUSG00000100608      2             1  MGI:5440224  OTTMUSG00000047316.1   \n",
       "ENSMUSG00000102011      2             1  MGI:5579636  OTTMUSG00000047349.1   \n",
       "ENSMUSG00000096178      2             7  MGI:5434193  OTTMUSG00000047352.1   \n",
       "ENSMUSG00000096768      2             8  MGI:6096131  OTTMUSG00000074820.2   \n",
       "\n",
       "                                  tag  \n",
       "ENSMUSG00000051951                NaN  \n",
       "ENSMUSG00000102343                NaN  \n",
       "ENSMUSG00000025900  overlapping_locus  \n",
       "ENSMUSG00000025902                NaN  \n",
       "ENSMUSG00000104238                NaN  \n",
       "...                               ...  \n",
       "ENSMUSG00000100574                NaN  \n",
       "ENSMUSG00000100608                NaN  \n",
       "ENSMUSG00000102011                NaN  \n",
       "ENSMUSG00000096178                NaN  \n",
       "ENSMUSG00000096768                NaN  \n",
       "\n",
       "[27651 rows x 22 columns]"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "use_gene_meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_status</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>level</th>\n",
       "      <th>gene_version</th>\n",
       "      <th>mgi_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>tag</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000051951</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3205901</td>\n",
       "      <td>3671498</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>MGI:3528744</td>\n",
       "      <td>OTTMUSG00000026353.2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000102343</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3905739</td>\n",
       "      <td>3986215</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:5610609</td>\n",
       "      <td>OTTMUSG00000049934.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000025900</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3999557</td>\n",
       "      <td>4409241</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>12</td>\n",
       "      <td>MGI:1341105</td>\n",
       "      <td>OTTMUSG00000049985.3</td>\n",
       "      <td>overlapping_locus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000025902</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4490931</td>\n",
       "      <td>4497354</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>13</td>\n",
       "      <td>MGI:107543</td>\n",
       "      <td>OTTMUSG00000050014.7</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000104238</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>4496551</td>\n",
       "      <td>4499558</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:5610815</td>\n",
       "      <td>OTTMUSG00000050024.2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000100574</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>89636206</td>\n",
       "      <td>89639438</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:5579380</td>\n",
       "      <td>OTTMUSG00000047315.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000100608</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>89713424</td>\n",
       "      <td>89745532</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:5440224</td>\n",
       "      <td>OTTMUSG00000047316.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000102011</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>89927120</td>\n",
       "      <td>89935013</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>MGI:5579636</td>\n",
       "      <td>OTTMUSG00000047349.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000096178</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>90401248</td>\n",
       "      <td>90433263</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>MGI:5434193</td>\n",
       "      <td>OTTMUSG00000047352.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000096768</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>90784738</td>\n",
       "      <td>90816465</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>MGI:6096131</td>\n",
       "      <td>OTTMUSG00000074820.2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>27651 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   chrom  source feature     start       end score strand  \\\n",
       "ENSMUSG00000051951  chr1  HAVANA    gene   3205901   3671498     .      -   \n",
       "ENSMUSG00000102343  chr1  HAVANA    gene   3905739   3986215     .      -   \n",
       "ENSMUSG00000025900  chr1  HAVANA    gene   3999557   4409241     .      -   \n",
       "ENSMUSG00000025902  chr1  HAVANA    gene   4490931   4497354     .      -   \n",
       "ENSMUSG00000104238  chr1  HAVANA    gene   4496551   4499558     .      +   \n",
       "...                  ...     ...     ...       ...       ...   ...    ...   \n",
       "ENSMUSG00000100574  chrY  HAVANA    gene  89636206  89639438     .      -   \n",
       "ENSMUSG00000100608  chrY  HAVANA    gene  89713424  89745532     .      -   \n",
       "ENSMUSG00000102011  chrY  HAVANA    gene  89927120  89935013     .      -   \n",
       "ENSMUSG00000096178  chrY  HAVANA    gene  90401248  90433263     .      +   \n",
       "ENSMUSG00000096768  chrY  HAVANA    gene  90784738  90816465     .      +   \n",
       "\n",
       "                   phase  transcript_id       gene_type  ...  transcript_type  \\\n",
       "ENSMUSG00000051951     .            NaN  protein_coding  ...              NaN   \n",
       "ENSMUSG00000102343     .            NaN          lncRNA  ...              NaN   \n",
       "ENSMUSG00000025900     .            NaN  protein_coding  ...              NaN   \n",
       "ENSMUSG00000025902     .            NaN  protein_coding  ...              NaN   \n",
       "ENSMUSG00000104238     .            NaN          lncRNA  ...              NaN   \n",
       "...                  ...            ...             ...  ...              ...   \n",
       "ENSMUSG00000100574     .            NaN          lncRNA  ...              NaN   \n",
       "ENSMUSG00000100608     .            NaN  protein_coding  ...              NaN   \n",
       "ENSMUSG00000102011     .            NaN          lncRNA  ...              NaN   \n",
       "ENSMUSG00000096178     .            NaN  protein_coding  ...              NaN   \n",
       "ENSMUSG00000096768     .            NaN          lncRNA  ...              NaN   \n",
       "\n",
       "                   transcript_status  transcript_name  exon_number  exon_id  \\\n",
       "ENSMUSG00000051951               NaN              NaN          NaN      NaN   \n",
       "ENSMUSG00000102343               NaN              NaN          NaN      NaN   \n",
       "ENSMUSG00000025900               NaN              NaN          NaN      NaN   \n",
       "ENSMUSG00000025902               NaN              NaN          NaN      NaN   \n",
       "ENSMUSG00000104238               NaN              NaN          NaN      NaN   \n",
       "...                              ...              ...          ...      ...   \n",
       "ENSMUSG00000100574               NaN              NaN          NaN      NaN   \n",
       "ENSMUSG00000100608               NaN              NaN          NaN      NaN   \n",
       "ENSMUSG00000102011               NaN              NaN          NaN      NaN   \n",
       "ENSMUSG00000096178               NaN              NaN          NaN      NaN   \n",
       "ENSMUSG00000096768               NaN              NaN          NaN      NaN   \n",
       "\n",
       "                    level  gene_version       mgi_id           havana_gene  \\\n",
       "ENSMUSG00000051951      2             5  MGI:3528744  OTTMUSG00000026353.2   \n",
       "ENSMUSG00000102343      2             1  MGI:5610609  OTTMUSG00000049934.1   \n",
       "ENSMUSG00000025900      2            12  MGI:1341105  OTTMUSG00000049985.3   \n",
       "ENSMUSG00000025902      2            13   MGI:107543  OTTMUSG00000050014.7   \n",
       "ENSMUSG00000104238      2             1  MGI:5610815  OTTMUSG00000050024.2   \n",
       "...                   ...           ...          ...                   ...   \n",
       "ENSMUSG00000100574      2             1  MGI:5579380  OTTMUSG00000047315.1   \n",
       "ENSMUSG00000100608      2             1  MGI:5440224  OTTMUSG00000047316.1   \n",
       "ENSMUSG00000102011      2             1  MGI:5579636  OTTMUSG00000047349.1   \n",
       "ENSMUSG00000096178      2             7  MGI:5434193  OTTMUSG00000047352.1   \n",
       "ENSMUSG00000096768      2             8  MGI:6096131  OTTMUSG00000074820.2   \n",
       "\n",
       "                                  tag  \n",
       "ENSMUSG00000051951                NaN  \n",
       "ENSMUSG00000102343                NaN  \n",
       "ENSMUSG00000025900  overlapping_locus  \n",
       "ENSMUSG00000025902                NaN  \n",
       "ENSMUSG00000104238                NaN  \n",
       "...                               ...  \n",
       "ENSMUSG00000100574                NaN  \n",
       "ENSMUSG00000100608                NaN  \n",
       "ENSMUSG00000102011                NaN  \n",
       "ENSMUSG00000096178                NaN  \n",
       "ENSMUSG00000096768                NaN  \n",
       "\n",
       "[27651 rows x 22 columns]"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bw_file = bw_files_mseq[0]\n",
    "\n",
    "cell_group = os.path.basename(bw_file).split('.')[0]\n",
    "cell_groups.append(cell_group)\n",
    "methtype = os.path.basename(bw_file).split('.')[1]\n",
    "methtypes.append(methtype)\n",
    "bw = pyBigWig.open(bw_file)\n",
    "\n",
    "# %%\n",
    "bw.isBigWig()\n",
    "\n",
    "# %%\n",
    "use_gene_meta\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Calculate the mean methylation level for each gene in each enhancer region\n",
    "# %%\n",
    "# stats =[]\n",
    "# CGN_values = []\n",
    "# matching_annotations = []\n",
    "\n",
    "# #for chrom, length in bb_file.chroms().items():\n",
    "# for index,row in use_gene_meta.iterrows():\n",
    "#     # Fetch all entries for the chromosome\n",
    "#     chrom = row['chrom']\n",
    "#     start = row['start']\n",
    "#     end = row['end']\n",
    "#     flaking = 2000\n",
    "\n",
    "#     entries = bb_file.entries(chrom, start-flaking, end+flaking)\n",
    "    \n",
    "#     # Filter entries for genes in the DataFrame\n",
    "#     if  entries is None:\n",
    "#         continue\n",
    "#     for entry in entries:\n",
    "#         start, end, name = entry[:3]  # Extract start, end, and name\n",
    "#         values = bw.values(chrom, start, end)\n",
    "#         arr_values = np.array(values)\n",
    "#         arr_values = np.nan_to_num(arr_values)\n",
    "#         stats.append(arr_values.mean())\n",
    "#         CGN_values.append(arr_values)\n",
    "#         matching_annotations.append(name)\n",
    "\n",
    "# maxlength = max([len(x) for x in CGN_values])\n",
    "# CGN_values = [np.pad(x,(0,maxlength-len(x))) for x in CGN_values]\n",
    "# enhancer_methylation = np.array(CGN_values)\n",
    "\n",
    "\n",
    "# if not os.path.exists(f'output/enhancer_meth/{cell_group}'):\n",
    "#     os.makedirs(f'output/enhancer_meth/{cell_group}')\n",
    "# plt.hist(stats, bins=100, color='skyblue', edgecolor='black')\n",
    "# # Add labels and title\n",
    "# plt.xlabel('Value')\n",
    "# plt.ylabel('Frequency')\n",
    "# plt.title('Histogram of Random Data')\n",
    "# plt.savefig(f'output/enhancer_meth/{cell_group}/{methtype}.png')\n",
    "# plt.close()\n",
    "\n",
    "# # value_array = np.array(CGN_values)\n",
    "# # np.save(f'output/enhancer_meth/{cell_group}/{methtype}.npy',value_array)\n",
    "# # stats = np.array(stats)\n",
    "# # np.save(f'output/enhancer_meth/{cell_group}/{methtype}_stats.npy',stats)\n",
    "# np.save(f'output/enhancer_meth/{cell_group}/{methtype}_enhancer_meth.npy',enhancer_methylation)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "def bb_to_bed(bb_file, bed_file):\n",
    "    chrom_sizes = bb_file.chroms()\n",
    "    \n",
    "    with open(bed_file, \"w\") as outfile:\n",
    "        for chrom, size in chrom_sizes.items():\n",
    "            entries = bb_file.entries(chrom, 0, size)\n",
    "            for entry in entries:\n",
    "                start, end, name = entry[:3]  # Extract start, end, and name\n",
    "                name_parts = name.split(\"\\t\")\n",
    "                if \"PLS\" in name_parts[6]:\n",
    "                    continue\n",
    "                else:\n",
    "                    value = name_parts[6]\n",
    "\n",
    "                outfile.write(f\"{chrom}\\t{start}\\t{end}\\t.\\t.\\t{value}\\n\")\n",
    "bb_to_bed(bb_file, f'output/convert_enhancer.bed')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "df_bed = pd.read_csv('output/convert_enhancer.bed',sep='\\t',header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'CTCF-only,CTCF-bound',\n",
       " 'DNase-H3K4me3',\n",
       " 'DNase-H3K4me3,CTCF-bound',\n",
       " 'dELS',\n",
       " 'dELS,CTCF-bound',\n",
       " 'pELS',\n",
       " 'pELS,CTCF-bound'}"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(df_bed.iloc[:,5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "encode = {'CTCF-only,CTCF-bound':'c',\n",
    " 'DNase-H3K4me3':'k',\n",
    " 'DNase-H3K4me3,CTCF-bound':'l',\n",
    " 'dELS':'d',\n",
    " 'dELS,CTCF-bound':'e',\n",
    " 'pELS':'p',\n",
    " 'pELS,CTCF-bound':\"q\"}\n",
    "\n",
    "df_bed.iloc[:,5] = df_bed.iloc[:,5].map(encode)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_bed.to_csv('output/convert_enhancer.bed',sep='\\t',header=None,index=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### They are mm10 files, convert them if needed \n",
    "#### https://genome.ucsc.edu/cgi-bin/hgLiftOver"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_m39_converted = pd.read_csv('output/enhancer_m39converted.bed',sep='\\t',header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "inverted_encode = {v: k for k, v in encode.items()}\n",
    "df_m39_converted.iloc[:,5] = df_m39_converted.iloc[:,5].map(inverted_encode)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_m39_converted.to_csv('output/enhancer_m39converted_renamed.bed',sep='\\t',header=None,index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pybedtools import BedTool\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "***** WARNING: File /data2st1/junyi/generegion_vM33/intergenic.bed has inconsistent naming convention for record:\n",
      "GL456210.1\t58883\t108389\t.\t.\tintergenic\n",
      "\n",
      "***** WARNING: File /data2st1/junyi/generegion_vM33/intergenic.bed has inconsistent naming convention for record:\n",
      "GL456210.1\t58883\t108389\t.\t.\tintergenic\n",
      "\n",
      "***** WARNING: File /data2st1/junyi/generegion_vM33/intergenic.bed has inconsistent naming convention for record:\n",
      "GL456210.1\t58883\t108389\t.\t.\tintergenic\n",
      "\n",
      "***** WARNING: File /data2st1/junyi/generegion_vM33/intergenic.bed has inconsistent naming convention for record:\n",
      "GL456210.1\t58883\t108389\t.\t.\tintergenic\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "error,chr1\t80405912\t80405911\t.\t.\tintergenic\n",
      "\n",
      "error,chr1\t80725132\t80724938\t.\t.\tintergenic\n",
      "\n",
      "error,chr1\t89855696\t89855677\t.\t.\tintergenic\n",
      "\n",
      "error,chr1\t106788913\t106788902\t.\t.\tintergenic\n",
      "\n",
      "error,chr1\t125488422\t125488331\t.\t.\tintergenic\n",
      "\n",
      "error,chr1\t131375470\t131375177\t.\t.\tintergenic\n",
      "\n",
      "error,chr1\t133092095\t133091947\t.\t.\tintergenic\n",
      "\n",
      "error,chr1\t172348441\t172348367\t.\t.\tintergenic\n",
      "\n",
      "error,chr1\t180591201\t180591061\t.\t.\tintergenic\n",
      "\n",
      "error,chr1\t193338019\t193337930\t.\t.\tintergenic\n",
      "\n",
      "error,chr2\t20673022\t20672896\t.\t.\tintergenic\n",
      "\n",
      "error,chr2\t30603365\t30603158\t.\t.\tintergenic\n",
      "\n",
      "error,chr2\t38559453\t38559428\t.\t.\tintergenic\n",
      "\n",
      "error,chr2\t71549697\t71549681\t.\t.\tintergenic\n",
      "\n",
      "error,chr2\t114684682\t114684573\t.\t.\tintergenic\n",
      "\n",
      "error,chr2\t118730924\t118730857\t.\t.\tintergenic\n",
      "\n",
      "error,chr2\t128275822\t128275642\t.\t.\tintergenic\n",
      "\n",
      "error,chr2\t147786438\t147786385\t.\t.\tintergenic\n",
      "\n",
      "error,chr2\t172819572\t172819492\t.\t.\tintergenic\n",
      "\n",
      "error,chr2\t173861196\t173861050\t.\t.\tintergenic\n",
      "\n",
      "error,chr2\t173918209\t173918100\t.\t.\tintergenic\n",
      "\n",
      "error,chr3\t9673884\t9673827\t.\t.\tintergenic\n",
      "\n",
      "error,chr3\t17843971\t17843891\t.\t.\tintergenic\n",
      "\n",
      "error,chr3\t51567901\t51567780\t.\t.\tintergenic\n",
      "\n",
      "error,chr3\t87420121\t87420081\t.\t.\tintergenic\n",
      "\n",
      "error,chr3\t88559244\t88559204\t.\t.\tintergenic\n",
      "\n",
      "error,chr3\t95641780\t95641458\t.\t.\tintergenic\n",
      "\n",
      "error,chr3\t95835989\t95835947\t.\t.\tintergenic\n",
      "\n",
      "error,chr3\t126390105\t126389950\t.\t.\tintergenic\n",
      "\n",
      "error,chr3\t126531176\t126531145\t.\t.\tintergenic\n",
      "\n",
      "error,chr3\t142199846\t142199738\t.\t.\tintergenic\n",
      "\n",
      "error,chr4\t4729727\t4729572\t.\t.\tintergenic\n",
      "\n",
      "error,chr4\t8939422\t8939231\t.\t.\tintergenic\n",
      "\n",
      "error,chr4\t43482416\t43482080\t.\t.\tintergenic\n",
      "\n",
      "error,chr4\t59260107\t59260050\t.\t.\tintergenic\n",
      "\n",
      "error,chr4\t76369156\t76369128\t.\t.\tintergenic\n",
      "\n",
      "error,chr4\t95040290\t95039976\t.\t.\tintergenic\n",
      "\n",
      "error,chr4\t106696465\t106696403\t.\t.\tintergenic\n",
      "\n",
      "error,chr4\t125039346\t125039331\t.\t.\tintergenic\n",
      "\n",
      "error,chr4\t132258045\t132257865\t.\t.\tintergenic\n",
      "\n",
      "error,chr4\t140675323\t140674976\t.\t.\tintergenic\n",
      "\n",
      "error,chr4\t141346269\t141346134\t.\t.\tintergenic\n",
      "\n",
      "error,chr4\t150910319\t150910189\t.\t.\tintergenic\n",
      "\n",
      "error,chr4\t152561257\t152561162\t.\t.\tintergenic\n",
      "\n",
      "error,chr4\t153457908\t153457677\t.\t.\tintergenic\n",
      "\n",
      "error,chr5\t8292967\t8292722\t.\t.\tintergenic\n",
      "\n",
      "error,chr5\t32084255\t32084040\t.\t.\tintergenic\n",
      "\n",
      "error,chr5\t53503747\t53503600\t.\t.\tintergenic\n",
      "\n",
      "error,chr5\t53623596\t53623493\t.\t.\tintergenic\n",
      "\n",
      "error,chr5\t64313866\t64313647\t.\t.\tintergenic\n",
      "\n",
      "error,chr5\t66681846\t66681529\t.\t.\tintergenic\n",
      "\n",
      "error,chr5\t81168018\t81167984\t.\t.\tintergenic\n",
      "\n",
      "error,chr5\t92275310\t92275106\t.\t.\tintergenic\n",
      "\n",
      "error,chr5\t97848666\t97848654\t.\t.\tintergenic\n",
      "\n",
      "error,chr5\t103573145\t103573057\t.\t.\tintergenic\n",
      "\n",
      "error,chr5\t105799307\t105799202\t.\t.\tintergenic\n",
      "\n",
      "error,chr5\t111638869\t111638796\t.\t.\tintergenic\n",
      "\n",
      "error,chr5\t116147470\t116147374\t.\t.\tintergenic\n",
      "\n",
      "error,chr5\t125094269\t125094216\t.\t.\tintergenic\n",
      "\n",
      "error,chr5\t125609094\t125608837\t.\t.\tintergenic\n",
      "\n",
      "error,chr5\t135835845\t135835743\t.\t.\tintergenic\n",
      "\n",
      "error,chr5\t140193186\t140193139\t.\t.\tintergenic\n",
      "\n",
      "error,chr5\t142889010\t142888869\t.\t.\tintergenic\n",
      "\n",
      "error,chr6\t28981574\t28981450\t.\t.\tintergenic\n",
      "\n",
      "error,chr6\t41093144\t41092940\t.\t.\tintergenic\n",
      "\n",
      "error,chr6\t41534786\t41534624\t.\t.\tintergenic\n",
      "\n",
      "error,chr6\t51306927\t51306780\t.\t.\tintergenic\n",
      "\n",
      "error,chr6\t54658004\t54657914\t.\t.\tintergenic\n",
      "\n",
      "error,chr6\t72930319\t72930218\t.\t.\tintergenic\n",
      "\n",
      "error,chr6\t88052416\t88052369\t.\t.\tintergenic\n",
      "\n",
      "error,chr6\t88129477\t88129362\t.\t.\tintergenic\n",
      "\n",
      "error,chr6\t101496676\t101496615\t.\t.\tintergenic\n",
      "\n",
      "error,chr6\t115782483\t115782465\t.\t.\tintergenic\n",
      "\n",
      "error,chr6\t121093793\t121093757\t.\t.\tintergenic\n",
      "\n",
      "error,chr6\t125409807\t125409626\t.\t.\tintergenic\n",
      "\n",
      "error,chr6\t136589083\t136589067\t.\t.\tintergenic\n",
      "\n",
      "error,chr6\t140546742\t140546643\t.\t.\tintergenic\n",
      "\n",
      "error,chr6\t145061521\t145061378\t.\t.\tintergenic\n",
      "\n",
      "error,chr6\t147908244\t147908077\t.\t.\tintergenic\n",
      "\n",
      "error,chr7\t16352896\t16352551\t.\t.\tintergenic\n",
      "\n",
      "error,chr7\t28085822\t28085570\t.\t.\tintergenic\n",
      "\n",
      "error,chr7\t44122976\t44122959\t.\t.\tintergenic\n",
      "\n",
      "error,chr7\t61440583\t61440266\t.\t.\tintergenic\n",
      "\n",
      "error,chr7\t63555282\t63555225\t.\t.\tintergenic\n",
      "\n",
      "error,chr7\t80510799\t80510667\t.\t.\tintergenic\n",
      "\n",
      "error,chr7\t96245943\t96245858\t.\t.\tintergenic\n",
      "\n",
      "error,chr7\t97373307\t97373212\t.\t.\tintergenic\n",
      "\n",
      "error,chr7\t99030345\t99030139\t.\t.\tintergenic\n",
      "\n",
      "error,chr7\t109117970\t109117947\t.\t.\tintergenic\n",
      "\n",
      "error,chr7\t110225550\t110225487\t.\t.\tintergenic\n",
      "\n",
      "error,chr7\t118079808\t118079717\t.\t.\tintergenic\n",
      "\n",
      "error,chr7\t132021846\t132021758\t.\t.\tintergenic\n",
      "\n",
      "error,chr7\t132314022\t132313810\t.\t.\tintergenic\n",
      "\n",
      "error,chr7\t136923767\t136923737\t.\t.\tintergenic\n",
      "\n",
      "error,chr7\t141392886\t141392806\t.\t.\tintergenic\n",
      "\n",
      "error,chr8\t12374045\t12373796\t.\t.\tintergenic\n",
      "\n",
      "error,chr8\t24996567\t24996445\t.\t.\tintergenic\n",
      "\n",
      "error,chr8\t26068629\t26068505\t.\t.\tintergenic\n",
      "\n",
      "error,chr8\t37016607\t37016480\t.\t.\tintergenic\n",
      "\n",
      "error,chr8\t41607199\t41607134\t.\t.\tintergenic\n",
      "\n",
      "error,chr8\t85388082\t85387959\t.\t.\tintergenic\n",
      "\n",
      "error,chr8\t89353307\t89353190\t.\t.\tintergenic\n",
      "\n",
      "error,chr8\t123291800\t123291670\t.\t.\tintergenic\n",
      "\n",
      "error,chr9\t35332559\t35332423\t.\t.\tintergenic\n",
      "\n",
      "error,chr9\t36733319\t36733159\t.\t.\tintergenic\n",
      "\n",
      "error,chr9\t45213433\t45213259\t.\t.\tintergenic\n",
      "\n",
      "error,chr9\t54441628\t54441429\t.\t.\tintergenic\n",
      "\n",
      "error,chr9\t58182843\t58182726\t.\t.\tintergenic\n",
      "\n",
      "error,chr9\t60241439\t60241424\t.\t.\tintergenic\n",
      "\n",
      "error,chr9\t70976440\t70976431\t.\t.\tintergenic\n",
      "\n",
      "error,chr9\t77743313\t77743311\t.\t.\tintergenic\n",
      "\n",
      "error,chr9\t103598586\t103598466\t.\t.\tintergenic\n",
      "\n",
      "error,chr10\t7505216\t7505136\t.\t.\tintergenic\n",
      "\n",
      "error,chr10\t12792405\t12792386\t.\t.\tintergenic\n",
      "\n",
      "error,chr10\t111690544\t111690425\t.\t.\tintergenic\n",
      "\n",
      "error,chr10\t116785813\t116785661\t.\t.\tintergenic\n",
      "\n",
      "error,chr10\t122223100\t122222796\t.\t.\tintergenic\n",
      "\n",
      "error,chr10\t126764450\t126764341\t.\t.\tintergenic\n",
      "\n",
      "error,chr11\t18955793\t18955606\t.\t.\tintergenic\n",
      "\n",
      "error,chr11\t32853172\t32853136\t.\t.\tintergenic\n",
      "\n",
      "error,chr11\t52250860\t52250846\t.\t.\tintergenic\n",
      "\n",
      "error,chr11\t55663020\t55662913\t.\t.\tintergenic\n",
      "\n",
      "error,chr11\t68321827\t68321507\t.\t.\tintergenic\n",
      "\n",
      "error,chr11\t69452597\t69452583\t.\t.\tintergenic\n",
      "\n",
      "error,chr11\t75351581\t75351274\t.\t.\tintergenic\n",
      "\n",
      "error,chr11\t81751746\t81751498\t.\t.\tintergenic\n",
      "\n",
      "error,chr11\t88609442\t88609353\t.\t.\tintergenic\n",
      "\n",
      "error,chr11\t96766602\t96766532\t.\t.\tintergenic\n",
      "\n",
      "error,chr11\t98337336\t98337219\t.\t.\tintergenic\n",
      "\n",
      "error,chr11\t108234185\t108234179\t.\t.\tintergenic\n",
      "\n",
      "error,chr11\t115353033\t115352870\t.\t.\tintergenic\n",
      "\n",
      "error,chr11\t117854560\t117854485\t.\t.\tintergenic\n",
      "\n",
      "error,chr11\t117867230\t117867034\t.\t.\tintergenic\n",
      "\n",
      "error,chr11\t118150770\t118150476\t.\t.\tintergenic\n",
      "\n",
      "error,chr11\t120117359\t120117213\t.\t.\tintergenic\n",
      "\n",
      "error,chr11\t120252433\t120252359\t.\t.\tintergenic\n",
      "\n",
      "error,chr11\t121693944\t121693932\t.\t.\tintergenic\n",
      "\n",
      "error,chr12\t8503564\t8503503\t.\t.\tintergenic\n",
      "\n",
      "error,chr12\t54363032\t54362949\t.\t.\tintergenic\n",
      "\n",
      "error,chr12\t69939684\t69939387\t.\t.\tintergenic\n",
      "\n",
      "error,chr12\t77019365\t77019233\t.\t.\tintergenic\n",
      "\n",
      "error,chr12\t102579098\t102578938\t.\t.\tintergenic\n",
      "\n",
      "error,chr12\t102639417\t102639275\t.\t.\tintergenic\n",
      "\n",
      "error,chr13\t23757982\t23757909\t.\t.\tintergenic\n",
      "\n",
      "error,chr13\t38150611\t38150580\t.\t.\tintergenic\n",
      "\n",
      "error,chr13\t44882993\t44882949\t.\t.\tintergenic\n",
      "\n",
      "error,chr13\t52127274\t52127249\t.\t.\tintergenic\n",
      "\n",
      "error,chr13\t56315555\t56315412\t.\t.\tintergenic\n",
      "\n",
      "error,chr13\t97647893\t97647711\t.\t.\tintergenic\n",
      "\n",
      "error,chr13\t101579537\t101579267\t.\t.\tintergenic\n",
      "\n",
      "error,chr14\t5650812\t5650539\t.\t.\tintergenic\n",
      "\n",
      "error,chr14\t6297400\t6297154\t.\t.\tintergenic\n",
      "\n",
      "error,chr14\t56060711\t56060600\t.\t.\tintergenic\n",
      "\n",
      "error,chr14\t67407040\t67407036\t.\t.\tintergenic\n",
      "\n",
      "error,chr14\t67470173\t67469885\t.\t.\tintergenic\n",
      "\n",
      "error,chr14\t77070291\t77070260\t.\t.\tintergenic\n",
      "\n",
      "error,chr14\t104185817\t104185531\t.\t.\tintergenic\n",
      "\n",
      "error,chr15\t5525797\t5525746\t.\t.\tintergenic\n",
      "\n",
      "error,chr15\t41210982\t41210826\t.\t.\tintergenic\n",
      "\n",
      "error,chr15\t82843745\t82843418\t.\t.\tintergenic\n",
      "\n",
      "error,chr15\t83368025\t83367969\t.\t.\tintergenic\n",
      "\n",
      "error,chr15\t99023861\t99023796\t.\t.\tintergenic\n",
      "\n",
      "error,chr15\t102008948\t102008847\t.\t.\tintergenic\n",
      "\n",
      "error,chr16\t8227377\t8227138\t.\t.\tintergenic\n",
      "\n",
      "error,chr16\t20270888\t20270683\t.\t.\tintergenic\n",
      "\n",
      "error,chr16\t60425233\t60425128\t.\t.\tintergenic\n",
      "\n",
      "error,chr17\t12337823\t12337590\t.\t.\tintergenic\n",
      "\n",
      "error,chr17\t26052169\t26052078\t.\t.\tintergenic\n",
      "\n",
      "error,chr17\t26882282\t26882204\t.\t.\tintergenic\n",
      "\n",
      "error,chr17\t27267317\t27267249\t.\t.\tintergenic\n",
      "\n",
      "error,chr17\t28220145\t28220038\t.\t.\tintergenic\n",
      "\n",
      "error,chr17\t28739191\t28738981\t.\t.\tintergenic\n",
      "\n",
      "error,chr17\t29115627\t29115365\t.\t.\tintergenic\n",
      "\n",
      "error,chr17\t35439202\t35439150\t.\t.\tintergenic\n",
      "\n",
      "error,chr17\t45941134\t45941129\t.\t.\tintergenic\n",
      "\n",
      "error,chr17\t56576476\t56576318\t.\t.\tintergenic\n",
      "\n",
      "error,chr17\t56818372\t56818247\t.\t.\tintergenic\n",
      "\n",
      "error,chr17\t64170234\t64170056\t.\t.\tintergenic\n",
      "\n",
      "error,chr17\t73143604\t73143447\t.\t.\tintergenic\n",
      "\n",
      "error,chr17\t84263765\t84263676\t.\t.\tintergenic\n",
      "\n",
      "error,chr18\t4590785\t4590762\t.\t.\tintergenic\n",
      "\n",
      "error,chr18\t5945291\t5945182\t.\t.\tintergenic\n",
      "\n",
      "error,chr18\t13088998\t13088908\t.\t.\tintergenic\n",
      "\n",
      "error,chr18\t34226508\t34226497\t.\t.\tintergenic\n",
      "\n",
      "error,chr18\t36021877\t36021821\t.\t.\tintergenic\n",
      "\n",
      "error,chr18\t38053182\t38052870\t.\t.\tintergenic\n",
      "\n",
      "error,chr18\t38319265\t38318966\t.\t.\tintergenic\n",
      "\n",
      "error,chr18\t38675944\t38675897\t.\t.\tintergenic\n",
      "\n",
      "error,chr18\t62310925\t62310886\t.\t.\tintergenic\n",
      "\n",
      "error,chr18\t77877662\t77877613\t.\t.\tintergenic\n",
      "\n",
      "error,chr19\t6806872\t6806576\t.\t.\tintergenic\n",
      "\n",
      "error,chr19\t7242602\t7242541\t.\t.\tintergenic\n",
      "\n",
      "error,chr19\t24533775\t24533652\t.\t.\tintergenic\n",
      "\n",
      "error,chr19\t32573269\t32573189\t.\t.\tintergenic\n",
      "\n",
      "error,chr19\t32937673\t32937530\t.\t.\tintergenic\n",
      "\n",
      "error,chr19\t33369355\t33369165\t.\t.\tintergenic\n",
      "\n",
      "error,chr19\t36325134\t36325013\t.\t.\tintergenic\n",
      "\n",
      "error,chr19\t36362226\t36362191\t.\t.\tintergenic\n",
      "\n",
      "error,chr19\t43516871\t43516750\t.\t.\tintergenic\n",
      "\n",
      "error,chr19\t45221301\t45221122\t.\t.\tintergenic\n",
      "\n",
      "error,chr19\t53117245\t53117114\t.\t.\tintergenic\n",
      "\n",
      "error,chr19\t58900159\t58900104\t.\t.\tintergenic\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# def merge_bed_files(bed_file1, bed_file2, output_file):\n",
    "#     # Load BED files\n",
    "#     bed1 = BedTool(bed_file1)\n",
    "#     bed2 = BedTool(bed_file2)\n",
    "\n",
    "#     # Intersect and split regions\n",
    "#     # Retain annotations from bed_file2 for overlapping regions\n",
    "#     merged = bed1.intersect(bed2, wa=True, wb=True).saveas(\"output/temp_intersect.bed\")\n",
    "\n",
    "#     # Split non-overlapping regions from bed_file1\n",
    "#     non_overlapping = bed1.subtract(bed2)#.saveas(\"output/temp_non_overlapping.bed\")\n",
    "\n",
    "#     start_tmp, end_tmp = 0, 0\n",
    "#     chrom_tmp = \"\"\n",
    "#     with open(\"output/temp_intersect.bed\", \"r\") as infile, open(\"output/temp_splited.bed\", \"w\") as outfile:\n",
    "#         lines = infile.readlines()\n",
    "#         for line in lines:\n",
    "#             fields = line.strip().split(\"\\t\")\n",
    "#             # Extract intergenic region\n",
    "#             chrom1, start1, end1, _, _, annotation1 = fields[:6]\n",
    "\n",
    "#             # Anew chromosome\n",
    "#             if chrom1 != chrom_tmp:\n",
    "#                 chrom_tmp = chrom1\n",
    "#                 start_tmp, end_tmp = 0, 0\n",
    "            \n",
    "#             start1, end1 = int(start1), int(end1)\n",
    "\n",
    "#             # Go to next region if start1 < start_tmp\n",
    "#             if start1 > end_tmp:\n",
    "#                 if start_tmp != 0 and end2<end_tmp:\n",
    "#                     if start_tmp > end_tmp:\n",
    "#                         print(f\"error,{chrom2}\\t{start_tmp}\\t{end_tmp}\\t.\\t.\\t{annotation1}\\n\")\n",
    "#                     elif start_tmp == end_tmp:\n",
    "#                         #print(f\"warning,{chrom2}\\t{start_tmp}\\t{end_tmp}\\t.\\t.\\t{annotation1}\\n\")\n",
    "#                         pass\n",
    "#                     else:\n",
    "#                         outfile.write(f\"{chrom2}\\t{start_tmp}\\t{end_tmp}\\t.\\t.\\t{annotation1}\\n\")\n",
    "#                 start_tmp, end_tmp = start1, end1\n",
    "\n",
    "#             # Extract enhancer region\n",
    "#             chrom2, start2, end2, _, _, annotation2 = fields[6:]\n",
    "#             start2, end2 = int(start2), int(end2)\n",
    "\n",
    "#             if start_tmp < start2:\n",
    "                \n",
    "#                 if start_tmp > start2-1:\n",
    "#                     print(f\"error,{chrom1}\\t{start_tmp}\\t{start2-1}\\t.\\t.\\t{annotation1}\\n\")\n",
    "#                 elif start_tmp == start2-1:\n",
    "#                     #print(f\"warning,{chrom1}\\t{start_tmp}\\t{start2-1}\\t.\\t.\\t{annotation1}\\n\")\n",
    "#                     pass\n",
    "#                 else:\n",
    "#                     outfile.write(f\"{chrom1}\\t{start_tmp}\\t{start2-1}\\t.\\t.\\t{annotation1}\\n\")\n",
    "\n",
    "#                 if end2 < start2:\n",
    "#                     print(f\"error,{chrom2}\\t{end2}\\t{end_tmp}\\t.\\t.\\t{annotation2}\\n\")\n",
    "#                 outfile.write(f\"{chrom2}\\t{start2}\\t{end2}\\t.\\t.\\t{annotation2}\\n\")\n",
    "#                 start_tmp = end2 +1\n",
    "    \n",
    "#     # bed_split = BedTool(\"output/temp_splited.bed\")\n",
    "#     # bed_split.cat(non_overlapping).saveas(output_file)\n",
    "    \n",
    "# merge_bed_files(\"/data2st1/junyi/generegion_vM33/intergenic.bed\", \"output/enhancer_m39converted.bed\", \"output/merged_output.bed\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "***** WARNING: File /data2st1/junyi/generegion_vM33/intergenic.bed has inconsistent naming convention for record:\n",
      "GL456210.1\t58883\t108389\t.\t.\tintergenic\n",
      "\n",
      "***** WARNING: File /data2st1/junyi/generegion_vM33/intergenic.bed has inconsistent naming convention for record:\n",
      "GL456210.1\t58883\t108389\t.\t.\tintergenic\n",
      "\n",
      "***** WARNING: File /data2st1/junyi/generegion_vM33/intergenic.bed has inconsistent naming convention for record:\n",
      "GL456210.1\t58883\t108389\t.\t.\tintergenic\n",
      "\n",
      "***** WARNING: File /data2st1/junyi/generegion_vM33/intergenic.bed has inconsistent naming convention for record:\n",
      "GL456210.1\t58883\t108389\t.\t.\tintergenic\n",
      "\n"
     ]
    }
   ],
   "source": [
    "def merge_bed_files(bed_file1, bed_file2, output_file):\n",
    "    # Load BED files\n",
    "    bed1 = BedTool(bed_file1)\n",
    "    bed2 = BedTool(bed_file2)\n",
    "\n",
    "    # Intersect and split regions\n",
    "    # Retain annotations from bed_file2 for overlapping regions\n",
    "    merged = bed1.intersect(bed2, wa=True, wb=True).saveas(\"output/temp_intersect.bed\")\n",
    "\n",
    "    # Split non-overlapping regions from bed_file1\n",
    "    non_overlapping = bed1.subtract(bed2).saveas(\"output/temp_non_overlapping.bed\")\n",
    "    \n",
    "    # Split non-overlapping regions from bed_file1\n",
    "    df_bed_1 = pd.read_csv('output/temp_intersect.bed',sep='\\t',header=None)\n",
    "    df_bed_2 = pd.read_csv('output/temp_non_overlapping.bed',sep='\\t',header=None)\n",
    "\n",
    "    def sort_key(chromosome):\n",
    "        if chromosome.startswith('chr'):\n",
    "            # Extract the numeric part for sorting\n",
    "            return (0, int(chromosome[3:])) if chromosome[3:].isdigit() else (1, chromosome)\n",
    "        return (2, chromosome)  # For any other cases\n",
    "    \n",
    "    df_bed_1=df_bed_1.iloc[:,6:]\n",
    "    df_bed_1.columns = ['chromosome','start','end','name','score','annotation']\n",
    "    df_bed_2.columns = ['chromosome','start','end','name','score','annotation']\n",
    "\n",
    "    df_cat = pd.concat([df_bed_1,df_bed_2]).sort_values(by=[\"start\",\"end\"])\n",
    "    df_cat['sort_key'] = df_cat['chromosome'].apply(sort_key)\n",
    "    df_sorted = df_cat.sort_values(by=['sort_key', 'start']).drop(columns='sort_key')\n",
    "    df_sorted=df_sorted[df_sorted.start!=df_sorted.end]\n",
    "    df_sorted.to_csv(output_file,sep='\\t',header=None,index=None)\n",
    "\n",
    "merge_bed_files(\"/data2st1/junyi/generegion_vM33/intergenic.bed\", \"output/enhancer_m39converted.bed\", \"output/merged_output.bed\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_bed_file(bed_file):\n",
    "    \"\"\"\n",
    "    Check if a BED file contains any regions where start > end.\n",
    "    \"\"\"\n",
    "    with open(bed_file, \"r\") as file:\n",
    "        for line_number, line in enumerate(file, start=1):\n",
    "            # Skip comment lines (if any)\n",
    "            if line.startswith(\"#\") or line.strip() == \"\":\n",
    "                continue\n",
    "            \n",
    "            # Split the line into fields\n",
    "            fields = line.strip().split(\"\\t\")\n",
    "            \n",
    "            # Ensure the line has at least 3 fields (chrom, start, end)\n",
    "            if len(fields) < 3:\n",
    "                print(f\"Warning: Line {line_number} has fewer than 3 fields: {line.strip()}\")\n",
    "                continue\n",
    "            \n",
    "            # Extract chrom, start, and end\n",
    "            chrom, start, end = fields[0], fields[1], fields[2]\n",
    "            \n",
    "            # Convert start and end to integers\n",
    "            try:\n",
    "                start = int(start)\n",
    "                end = int(end)\n",
    "            except ValueError:\n",
    "                print(f\"Warning: Line {line_number} has non-integer start/end values: {line.strip()}\")\n",
    "                continue\n",
    "            \n",
    "            # Check if start > end\n",
    "            if start > end:\n",
    "                print(f\"Error: Line {line_number} has start > end: {line.strip()}\")\n",
    "            \n",
    "            # Optional: Check if start == end (zero-length regions)\n",
    "            if start == end:\n",
    "                print(f\"Warning: Line {line_number} has start == end (zero-length region): {line.strip()}\")\n",
    "\n",
    "# Example usage\n",
    "check_bed_file(\"output/merged_output.bed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<BedTool(output/merged_output.bed)>"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bed1 = BedTool(\"output/temp_splited.bed\")\n",
    "bed2 = BedTool(\"output/output/temp_splited.bed\")\n",
    "\n",
    "bed1.cat(bed2).merge().saveas(\"output/merged_output.bed\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "allcools",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.20"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
